{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EXONS Cas9 vs Cpf1 Chopchop\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-07-09T19:04:15.143416Z",
     "start_time": "2019-07-09T19:04:14.216887Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.11.5 | packaged by conda-forge | (main, Aug 27 2023, 03:34:09) [GCC 12.3.0]\n",
      "numpy 1.26.0\n",
      "scipy 1.12.0\n",
      "pandas 2.1.1\n",
      "scikit-allel 1.3.7\n",
      "zarr 2.16.1\n",
      "statsmodels 0.14.1\n"
     ]
    }
   ],
   "source": [
    "import sys; print(sys.version)\n",
    "import os\n",
    "import glob\n",
    "import subprocess\n",
    "import multiprocessing\n",
    "import io\n",
    "from collections import OrderedDict\n",
    "import json\n",
    "import pickle\n",
    "\n",
    "import numpy as np; print('numpy', np.__version__)\n",
    "import scipy; print('scipy', scipy.__version__)\n",
    "import pandas as pd; print('pandas',pd.__version__)\n",
    "import allel; print('scikit-allel', allel.__version__)\n",
    "import zarr; print('zarr', zarr.__version__)\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib \n",
    "\n",
    "import statsmodels; print('statsmodels', statsmodels.__version__)\n",
    "import statsmodels.api as sm\n",
    "\n",
    "from IPython.display import display, HTML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-07-09T19:04:15.150021Z",
     "start_time": "2019-07-09T19:04:15.144976Z"
    }
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "mpl.rcParams['figure.facecolor'] = '#BBBBBB'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bit of code to showing all a dataframe\n",
    "from IPython.display import display, HTML\n",
    "def display_all(df):\n",
    "    with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', None):\n",
    "        display(df)\n",
    "def display_all_cols(df, max_rows=60):\n",
    "    with pd.option_context('display.max_rows', max_rows, 'display.max_columns', None, 'display.width', None):\n",
    "        display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Settings/Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "SETNAME = 'AAEL_Af'\n",
    "\n",
    "if SETNAME == 'AgamGF':\n",
    "    ## AgamGF ##########\n",
    "    GENOME = \"AgamP4.12\"\n",
    "    GENE_PREFIX = \"AGAP\"\n",
    "    SAMPLE_LIST_FN = None\n",
    "    CHROMS = ['2R','2L','3R','3L','X']\n",
    "    GFFFN = \"data/Anopheles-gambiae-PEST_BASEFEATURES_AgamP4.12.gff3\"\n",
    "    ZVCF = 'data/YLAgamGF+Ag1000G_pfilt.vcf.gz.zarr'\n",
    "    CHOPCHOP_OUT_DIR = {'Cas9':\"Cas9_all_transcripts_run\",\n",
    "                        'Cpf1':\"Cpf1_all_transcripts_run\"}\n",
    "    TRANSFN = \"data/AgamP4.12.transcript_list\"\n",
    "\n",
    "elif SETNAME == 'AAEL':\n",
    "    ## AaegL5 #########\n",
    "    GENOME = \"AaegL5.1\"\n",
    "    GENE_PREFIX = \"AAEL\"\n",
    "    SAMPLE_LIST_FN = None\n",
    "    CHROMS = ['1','2','3']\n",
    "    GFFFN = \"data/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.1.gff3\"\n",
    "    ZVCF = 'data/AaegL5_2023-03.vcf.gz.zarr'\n",
    "    CHOPCHOP_OUT_DIR = {'Cas9':\"AaegL5_Cas9_all_transcripts_run\",\n",
    "                        'Cpf1':\"AaegL5_Cpf1_all_transcripts_run\"}\n",
    "    TRANSFN = \"data/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.1.transcript_list\"\n",
    "\n",
    "elif SETNAME == 'AAEL_Am':\n",
    "    ## AaegL5 American samples #########\n",
    "    GENOME = \"AaegL5.1\"\n",
    "    GENE_PREFIX = \"AAEL\"\n",
    "    SAMPLE_LIST_FN = \"AAEG_sample_group_Am.csv\"\n",
    "    CHROMS = ['1','2','3']\n",
    "    GFFFN = \"data/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.1.gff3\"\n",
    "    ZVCF = 'data/AaegL5_2023-03.vcf.gz.zarr'\n",
    "    CHOPCHOP_OUT_DIR = {'Cas9':\"AaegL5_Cas9_all_transcripts_run\",\n",
    "                        'Cpf1':\"AaegL5_Cpf1_all_transcripts_run\"}\n",
    "    TRANSFN = \"data/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.1.transcript_list\"\n",
    "    \n",
    "elif SETNAME == 'AAEL_Af':\n",
    "    ## AaegL5 American samples #########\n",
    "    GENOME = \"AaegL5.1\"\n",
    "    GENE_PREFIX = \"AAEL\"\n",
    "    SAMPLE_LIST_FN = \"AAEG_sample_group_Af.csv\"\n",
    "    CHROMS = ['1','2','3']\n",
    "    GFFFN = \"data/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.1.gff3\"\n",
    "    ZVCF = 'data/AaegL5_2023-03.vcf.gz.zarr'\n",
    "    CHOPCHOP_OUT_DIR = {'Cas9':\"AaegL5_Cas9_all_transcripts_run\",\n",
    "                        'Cpf1':\"AaegL5_Cpf1_all_transcripts_run\"}\n",
    "    TRANSFN = \"data/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.1.transcript_list\"\n",
    "    \n",
    "else:\n",
    "    assert False, f\"UNKNOWN SETNAME '{SETNAME}'\"\n",
    "    \n",
    "    \n",
    "##############################################################\n",
    "\n",
    "MIN_GC = 40 # minimum allowable GC percentage (inclusive)\n",
    "MAX_GC = 80 # maximum allowable GC percentage (inclusive)\n",
    "MIN_EFFICIENCY = 0.50 \n",
    "MAX_OFFTARGET_HITS = [0,0,0,0] # maximum allowable total offtarget hits (MM0, MM1, MM2, MM3); None=don't filter\n",
    "\n",
    "SUBSAMPLE_N = 0\n",
    "IGNORE_ONE_CALL_VARIANTS = False\n",
    "\n",
    "AWK_EXEC = '/usr/bin/awk'\n",
    "\n",
    "###################################################################\n",
    "# Cas9 gRNA: gRNA sequence (20 nt) in blue and PAM in red (N being any nucleotide)\n",
    "# The cut site is 3 nucleotides in taking only the gRNA sequence as a reference or 6 nucleotides in if you count the 3 nucleotides for the PAM site (in red)\n",
    "#\n",
    "# 5’ GGCGTATATATCGCGTAGTC NGG 3’\n",
    "# 5’ GGCGTATATATCGCGTA-(cut site)-GTC NGG 3’\n",
    "#\n",
    "# Cas12a gRNA: gRNA sequence (23 nt) in blue and PAM in red (N being any nucleotide)\n",
    "# The cut site for the Cas12a is staggered. So, it makes two simultaneous cuts at two different positions (cut site 1 and cut site 2)\n",
    "# Cut site one is 18–19 bases from the 3′ end of the PAM. The cut on the opposite strand of the DNA is 23 bases from the PAM, causing a 5′ overhang of 4–5 bases. \n",
    "#\n",
    "# 5’ TTTN GGCGTATATATCGCGTAG-(cut site)-TCGGG 3’ – “cut site 1”\n",
    "# 5’ TTTN GGCGTATATATCGCGTAGTCGGG-(cut site) 3’ – “cut site 2”\n",
    "######\n",
    "# Cas9 (gRNA is 23bp):\n",
    "# 1 \t34 \tTTATCCTTGTTTCAAAC TATAGG \t2L:157348 \t+  (loc+17)\n",
    "#                             ^ \n",
    "# 3 \t61 \tTATTATGTATTAAATAA TATTGG \t2L:157466 \t-  (loc+5)\n",
    "#        rc CCAATA TTATTTAATACATAATA\n",
    "#                ^\n",
    "# Cpf1 (gRNA is 28bp):\n",
    "# 3 \t37 \tTTTCAAACTATAGGTGCGTGAAA ATCG C \t2L:157357 \t+ \t\n",
    "#                                   ^    ^                (loc+23, loc+27)\n",
    "# 1 \t16 \tTTTCACGCACCTATAGTTTGAAA CAAG G \t2L:157352 \t- \t\n",
    "#        rc C CTTG TTTCAAACTATAGGTGCGTGAAA\n",
    "#           ^    ^                                        (loc+0, loc+4)\n",
    "CUT_OFFSET_gRNApos_CDSpos = {'Cas9':17, 'Cpf1':23}\n",
    "CUT_OFFSET_gRNApos_CDSneg = {'Cas9':17, 'Cpf1':27}\n",
    "CUT_OFFSET_gRNAneg_CDSpos = {'Cas9':5,  'Cpf1':4}\n",
    "CUT_OFFSET_gRNAneg_CDSneg = {'Cas9':5,  'Cpf1':0}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-07-09T19:04:41.701381Z",
     "start_time": "2019-07-09T19:04:41.698523Z"
    }
   },
   "outputs": [],
   "source": [
    "results_dict = OrderedDict()  # to collect results for final display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the trascript list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-07-09T19:05:09.356168Z",
     "start_time": "2019-07-09T19:05:09.290196Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAEL008159-RA</td>\n",
       "      <td>AAEL008159</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAEL026895-RA</td>\n",
       "      <td>AAEL026895</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAEL020319-RA</td>\n",
       "      <td>AAEL020319</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAEL025965-RB</td>\n",
       "      <td>AAEL025965</td>\n",
       "      <td>RB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAEL025965-RA</td>\n",
       "      <td>AAEL025965</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29330</th>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29331</th>\n",
       "      <td>AAEL012106-RH</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>RH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29332</th>\n",
       "      <td>AAEL012106-RG</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>RG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29333</th>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29334</th>\n",
       "      <td>AAEL012102-RB</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29335 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id        gene splice_id\n",
       "0      AAEL008159-RA  AAEL008159        RA\n",
       "1      AAEL026895-RA  AAEL026895        RA\n",
       "2      AAEL020319-RA  AAEL020319        RA\n",
       "3      AAEL025965-RB  AAEL025965        RB\n",
       "4      AAEL025965-RA  AAEL025965        RA\n",
       "...              ...         ...       ...\n",
       "29330  AAEL012106-RA  AAEL012106        RA\n",
       "29331  AAEL012106-RH  AAEL012106        RH\n",
       "29332  AAEL012106-RG  AAEL012106        RG\n",
       "29333  AAEL012102-RA  AAEL012102        RA\n",
       "29334  AAEL012102-RB  AAEL012102        RB\n",
       "\n",
       "[29335 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translist = pd.read_csv(TRANSFN, header=None, names=['transcript_id'])\n",
    "t = translist['transcript_id'].str.rsplit('-', n=1, expand=True)\n",
    "t.columns = ['gene', 'splice_id']\n",
    "translist = pd.concat((translist,t), axis=1)\n",
    "translist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of genes in gtf: 14626\n",
      "total number of pseudogenes in gtf: 382\n",
      "CDSs not matching a transcript from translist: 0\n",
      "translist entries without a CDS (presumably pseudogenes): 1019\n",
      "CDSs not in CHROMS ['1', '2', '3']: 4067\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>31666</td>\n",
       "      <td>31761</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>31666</td>\n",
       "      <td>31761</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PB;Parent=AAEL012102-RB;protein_...</td>\n",
       "      <td>AAEL012102-RB</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>31832</td>\n",
       "      <td>32404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>31832</td>\n",
       "      <td>32404</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PB;Parent=AAEL012102-RB;protein_...</td>\n",
       "      <td>AAEL012102-RB</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>32495</td>\n",
       "      <td>32673</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168730</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>409622594</td>\n",
       "      <td>409622821</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL010311-PA;Parent=AAEL010311-RA;protein_...</td>\n",
       "      <td>AAEL010311-RA</td>\n",
       "      <td>AAEL010311</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168731</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>409622594</td>\n",
       "      <td>409622821</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL010311-PB;Parent=AAEL010311-RB;protein_...</td>\n",
       "      <td>AAEL010311-RB</td>\n",
       "      <td>AAEL010311</td>\n",
       "      <td>RB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168732</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>409688525</td>\n",
       "      <td>409688690</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168733</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>409695468</td>\n",
       "      <td>409696934</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168734</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>409696991</td>\n",
       "      <td>409697244</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>RA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168735 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqid      source type      start        end score strand phase  \\\n",
       "0          1  VectorBase  CDS      31666      31761     .      +     0   \n",
       "1          1  VectorBase  CDS      31666      31761     .      +     0   \n",
       "2          1  VectorBase  CDS      31832      32404     .      +     0   \n",
       "3          1  VectorBase  CDS      31832      32404     .      +     0   \n",
       "4          1  VectorBase  CDS      32495      32673     .      +     0   \n",
       "...      ...         ...  ...        ...        ...   ...    ...   ...   \n",
       "168730     3  VectorBase  CDS  409622594  409622821     .      -     0   \n",
       "168731     3  VectorBase  CDS  409622594  409622821     .      -     0   \n",
       "168732     3  VectorBase  CDS  409688525  409688690     .      +     0   \n",
       "168733     3  VectorBase  CDS  409695468  409696934     .      +     2   \n",
       "168734     3  VectorBase  CDS  409696991  409697244     .      +     2   \n",
       "\n",
       "                                               attributes  transcript_id  \\\n",
       "0       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "1       ID=AAEL012102-PB;Parent=AAEL012102-RB;protein_...  AAEL012102-RB   \n",
       "2       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "3       ID=AAEL012102-PB;Parent=AAEL012102-RB;protein_...  AAEL012102-RB   \n",
       "4       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "...                                                   ...            ...   \n",
       "168730  ID=AAEL010311-PA;Parent=AAEL010311-RA;protein_...  AAEL010311-RA   \n",
       "168731  ID=AAEL010311-PB;Parent=AAEL010311-RB;protein_...  AAEL010311-RB   \n",
       "168732  ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...  AAEL010318-RA   \n",
       "168733  ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...  AAEL010318-RA   \n",
       "168734  ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...  AAEL010318-RA   \n",
       "\n",
       "              gene splice_id  \n",
       "0       AAEL012102        RA  \n",
       "1       AAEL012102        RB  \n",
       "2       AAEL012102        RA  \n",
       "3       AAEL012102        RB  \n",
       "4       AAEL012102        RA  \n",
       "...            ...       ...  \n",
       "168730  AAEL010311        RA  \n",
       "168731  AAEL010311        RB  \n",
       "168732  AAEL010318        RA  \n",
       "168733  AAEL010318        RA  \n",
       "168734  AAEL010318        RA  \n",
       "\n",
       "[168735 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#\n",
    "# Get the table of CDSs from the gff3 file\n",
    "#\n",
    "\n",
    "t = translist.copy(deep=True)\n",
    "\n",
    "assert t.shape[0] == t['transcript_id'].nunique(), \"trascript list should have no duplicte transcript_ids\"\n",
    "\n",
    "t.set_index('transcript_id', inplace=True)\n",
    "\n",
    "gffdf = pd.read_csv(GFFFN, sep='\\t', comment='#', header=None,\n",
    "            names=['seqid',\n",
    "                   'source',\n",
    "                   'type',\n",
    "                   'start',\n",
    "                   'end',\n",
    "                   'score',\n",
    "                   'strand',\n",
    "                   'phase',\n",
    "                   'attributes'],\n",
    "                dtype={'seqid':str})\n",
    "\n",
    "# total number of genes (just accounting / checking)\n",
    "# note: assumes one entry per gene & pseudogene in gff\n",
    "tmp = (gffdf['type'] == 'gene').sum()\n",
    "print(\"total number of genes in gtf:\", tmp)\n",
    "results_dict['total genes in gtf'] = tmp\n",
    "tmp2 = (gffdf['type'] == 'pseudogene').sum()\n",
    "print(\"total number of pseudogenes in gtf:\", tmp2)\n",
    "# results_dict['total pseudogenes in gtf'] = tmp2\n",
    "assert tmp+tmp2 == t['gene'].nunique(), \"number of genes+pseudogenes in gtf should equal number of genes in gene_table\"\n",
    "\n",
    "# just look at CDS entries\n",
    "cds = gffdf.loc[gffdf['type']=='CDS' ,:].copy(deep=True)\n",
    "# add transcript_id to Exon entries\n",
    "cds['transcript_id'] = cds['attributes'].apply(\n",
    "    lambda x: [x3[1].strip(\"'\") for x3 in [x2.split('=') for x2 in x.split(';')] if x3[0] == 'Parent' ][0])\n",
    "\n",
    "# join with the transcript list to get the gene_ids\n",
    "num_CDSs = cds.shape[0]\n",
    "cds = cds.join(translist.set_index('transcript_id'), on='transcript_id', how='outer')\n",
    "\n",
    "print(f\"CDSs not matching a transcript from translist: {(cds['gene'].isna()).sum()}\") # should be 0\n",
    "assert cds['gene'].isna().sum()==0, \"CDS not matching a transcript in translist\"\n",
    "\n",
    "print(f\"translist entries without a CDS (presumably pseudogenes): {(cds['seqid'].isna()).sum()}\")\n",
    "# drop the non-matching translist entries\n",
    "cds = cds.loc[~cds['seqid'].isna(),:]\n",
    "\n",
    "# cast positions to ints\n",
    "cds['start'] = cds['start'].astype(int)\n",
    "cds['end'] = cds['end'].astype(int)\n",
    "\n",
    "# Only take CDSs in the CRHOMS of interest\n",
    "tmp = cds.shape[0]\n",
    "cds = cds[cds['seqid'].isin(CHROMS)]\n",
    "print(f'CDSs not in CHROMS {CHROMS}: {tmp-cds.shape[0]}')\n",
    "\n",
    "cds.sort_values(['seqid','start','splice_id'], inplace=True)\n",
    "cds.reset_index(drop=True, inplace=True)\n",
    "\n",
    "cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DROPPING 79 CDSs where start==end\n",
      "number of CDS which are simple splice variants (same start,end,strand,gene): 103728\n",
      "DROPPING 103728 simple spliace variants\n",
      "number of same-gene overlap groups with multiple CDSs: 2667\n",
      "number of same-gene overlapping CDSs: 5684\n",
      "number of same-gene groups with CDSs of different lengths: 40203\n",
      "** number of protein coding genes: 13601\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>31832.0</td>\n",
       "      <td>32404.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>573.0</td>\n",
       "      <td>573.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>32732.0</td>\n",
       "      <td>33596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>865.0</td>\n",
       "      <td>865.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>41767.0</td>\n",
       "      <td>41958.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>44349.0</td>\n",
       "      <td>44471.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>RA</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>123.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168728</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL010311-PA;Parent=AAEL010311-RA;protein_...</td>\n",
       "      <td>AAEL010311-RA</td>\n",
       "      <td>AAEL010311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61907</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168730</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL010311-PA;Parent=AAEL010311-RA;protein_...</td>\n",
       "      <td>AAEL010311-RA</td>\n",
       "      <td>AAEL010311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61908</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168732</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61909</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168733</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168734</th>\n",
       "      <td>3</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61911</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61911 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqid      source type    start      end score strand phase  \\\n",
       "0          1  VectorBase  CDS      NaN      NaN     .      +     0   \n",
       "2          1  VectorBase  CDS  31832.0  32404.0     .      +     0   \n",
       "4          1  VectorBase  CDS  32732.0  33596.0     .      +     0   \n",
       "6          1  VectorBase  CDS  41767.0  41958.0     .      +     1   \n",
       "8          1  VectorBase  CDS  44349.0  44471.0     .      +     0   \n",
       "...      ...         ...  ...      ...      ...   ...    ...   ...   \n",
       "168728     3  VectorBase  CDS      NaN      NaN     .      -     0   \n",
       "168730     3  VectorBase  CDS      NaN      NaN     .      -     0   \n",
       "168732     3  VectorBase  CDS      NaN      NaN     .      +     0   \n",
       "168733     3  VectorBase  CDS      NaN      NaN     .      +     2   \n",
       "168734     3  VectorBase  CDS      NaN      NaN     .      +     2   \n",
       "\n",
       "                                               attributes  transcript_id  \\\n",
       "0       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "2       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "4       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "6       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "8       ID=AAEL012102-PA;Parent=AAEL012102-RA;protein_...  AAEL012102-RA   \n",
       "...                                                   ...            ...   \n",
       "168728  ID=AAEL010311-PA;Parent=AAEL010311-RA;protein_...  AAEL010311-RA   \n",
       "168730  ID=AAEL010311-PA;Parent=AAEL010311-RA;protein_...  AAEL010311-RA   \n",
       "168732  ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...  AAEL010318-RA   \n",
       "168733  ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...  AAEL010318-RA   \n",
       "168734  ID=AAEL010318-PA;Parent=AAEL010318-RA;protein_...  AAEL010318-RA   \n",
       "\n",
       "              gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \n",
       "0       AAEL012102       NaN              1      NaN       NaN          NaN  \n",
       "2       AAEL012102        RA              2      1.0     573.0        573.0  \n",
       "4       AAEL012102        RA              3      1.0     865.0        865.0  \n",
       "6       AAEL012102        RA              4      1.0     192.0        192.0  \n",
       "8       AAEL012102        RA              5      1.0     123.0        123.0  \n",
       "...            ...       ...            ...      ...       ...          ...  \n",
       "168728  AAEL010311       NaN          61907      NaN       NaN          NaN  \n",
       "168730  AAEL010311       NaN          61908      NaN       NaN          NaN  \n",
       "168732  AAEL010318       NaN          61909      NaN       NaN          NaN  \n",
       "168733  AAEL010318       NaN          61910      NaN       NaN          NaN  \n",
       "168734  AAEL010318       NaN          61911      NaN       NaN          NaN  \n",
       "\n",
       "[61911 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop CDSs which have length 1 (start==end)\n",
    "print(f\"DROPPING {(cds['start'] == cds['end']).sum()} CDSs where start==end\")\n",
    "cds = cds[cds['start'] != cds['end']].copy()\n",
    "\n",
    "\n",
    "# look for overlapping CDS with the SAME GENE\n",
    "print(f\"number of CDS which are simple splice variants (same start,end,strand,gene): {cds.duplicated(['seqid','start','end','strand','gene'], keep='first').sum()}\")\n",
    "\n",
    "# drop simple splice variants\n",
    "tmp = cds.shape[0]\n",
    "cds.drop_duplicates(['seqid','start','end','strand','gene'], keep='first', inplace=True)\n",
    "print(f\"DROPPING {tmp-cds.shape[0]} simple spliace variants\")\n",
    "\n",
    "\n",
    "\n",
    "# overlap groups with same gene\n",
    "overlap = ((cds['start']<=cds['end'].shift(1)) & (cds['seqid']==cds['seqid'].shift(1)) & (cds['gene']==cds['gene'].shift(1)))\n",
    "cds['overlap_group'] = (~overlap).cumsum()\n",
    "gb = cds.groupby('overlap_group', dropna=False)\n",
    "\n",
    "assert (gb['gene'].nunique() == 1).all(), \"all groups should contain only the same gene\"\n",
    "\n",
    "print(f\"number of same-gene overlap groups with multiple CDSs: {(gb['gene'].count()>1).sum()}\")\n",
    "print(f\"number of same-gene overlapping CDSs: {gb['gene'].count()[gb['gene'].count()>1].sum()}\")\n",
    "\n",
    "foo = gb.nth(0).copy()\n",
    "foo['start'] = gb['start'].min()\n",
    "foo['end'] = gb['end'].max()\n",
    "foo['CDS cnt'] = gb.size()\n",
    "foo['splice_id'] = gb['splice_id'].agg(lambda x: ','.join(x))\n",
    "\n",
    "foo['full_len'] = gb['end'].max()-gb['start'].min()+1\n",
    "foo['overlap_len'] = gb['end'].min()-gb['start'].max()+1\n",
    "\n",
    "print(f\"number of same-gene groups with CDSs of different lengths: {foo[foo['overlap_len'] != foo['full_len']].shape[0]}\")\n",
    "\n",
    "print(\"** number of protein coding genes:\", cds['gene'].nunique())\n",
    "results_dict['number of protein colding genes'] = cds['gene'].nunique()\n",
    "\n",
    "cds_grps = foo\n",
    "cds_grps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of overlap groups with multiple CDSs: 2681\n",
      "number of overlapping CDSs: 5714\n",
      "** total coding length [bp]: 23255031\n",
      "total coding len possibly multiple counting non-same-gene overlaps: 8370802.0\n"
     ]
    }
   ],
   "source": [
    "# computing the cumulatilve length (bp) of the CDSs accounting for overlaps\n",
    "overlap = ((cds['start']<=cds['end'].shift(1)) & (cds['seqid']==cds['seqid'].shift(1)))\n",
    "cds['overlap_group_ignoring_gene'] = (~overlap).cumsum()\n",
    "\n",
    "gb = cds.groupby('overlap_group_ignoring_gene', dropna=False)\n",
    "\n",
    "print(f\"number of overlap groups with multiple CDSs: {(gb['gene'].count()>1).sum()}\")\n",
    "print(f\"number of overlapping CDSs: {gb['gene'].count()[gb['gene'].count()>1].sum()}\")\n",
    "\n",
    "total_coding_len = (gb['end'].max()-gb['start'].min()+1).sum()\n",
    "print(f\"** total coding length [bp]: {total_coding_len}\")\n",
    "results_dict['total coding length [bp]'] = total_coding_len\n",
    "\n",
    "print(f\"total coding len possibly multiple counting non-same-gene overlaps: {cds_grps['full_len'].sum()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of overlaps involving diferent genes: 10\n",
      "number of CDSs with overlaps involving different genes: 37\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5770858.0</td>\n",
       "      <td>5771183.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL011377-PA;Parent=AAEL011377-RA;protein_...</td>\n",
       "      <td>AAEL011377-RA</td>\n",
       "      <td>AAEL011377</td>\n",
       "      <td>RB,RA,RE</td>\n",
       "      <td>128</td>\n",
       "      <td>3.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5771094.0</td>\n",
       "      <td>5771183.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL011381-PA;Parent=AAEL011381-RA;protein_...</td>\n",
       "      <td>AAEL011381-RA</td>\n",
       "      <td>AAEL011381</td>\n",
       "      <td>RE</td>\n",
       "      <td>129</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqid      source type      start        end score strand phase  \\\n",
       "306     1  VectorBase  CDS  5770858.0  5771183.0     .      -     0   \n",
       "307     1  VectorBase  CDS  5771094.0  5771183.0     .      -     0   \n",
       "\n",
       "                                            attributes  transcript_id  \\\n",
       "306  ID=AAEL011377-PA;Parent=AAEL011377-RA;protein_...  AAEL011377-RA   \n",
       "307  ID=AAEL011381-PA;Parent=AAEL011381-RA;protein_...  AAEL011381-RA   \n",
       "\n",
       "           gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "306  AAEL011377  RB,RA,RE            128      3.0     326.0         20.0   \n",
       "307  AAEL011381        RE            129      1.0      90.0         90.0   \n",
       "\n",
       "     overlap_group_ignoring_gene  \n",
       "306                          128  \n",
       "307                          128  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 326.0\n",
      "overlap len: 90.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17935465.0</td>\n",
       "      <td>17936815.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL000242-PD;Parent=AAEL000242-RD;protein_...</td>\n",
       "      <td>AAEL000242-RD</td>\n",
       "      <td>AAEL000242</td>\n",
       "      <td>RA,RB</td>\n",
       "      <td>319</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1351.0</td>\n",
       "      <td>1182.0</td>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17936704.0</td>\n",
       "      <td>17936815.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL000249-PA;Parent=AAEL000249-RA;protein_...</td>\n",
       "      <td>AAEL000249-RA</td>\n",
       "      <td>AAEL000249</td>\n",
       "      <td>RB</td>\n",
       "      <td>320</td>\n",
       "      <td>1.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqid      source type       start         end score strand phase  \\\n",
       "864     1  VectorBase  CDS  17935465.0  17936815.0     .      -     0   \n",
       "865     1  VectorBase  CDS  17936704.0  17936815.0     .      +     0   \n",
       "\n",
       "                                            attributes  transcript_id  \\\n",
       "864  ID=AAEL000242-PD;Parent=AAEL000242-RD;protein_...  AAEL000242-RD   \n",
       "865  ID=AAEL000249-PA;Parent=AAEL000249-RA;protein_...  AAEL000249-RA   \n",
       "\n",
       "           gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "864  AAEL000242     RA,RB            319      2.0    1351.0       1182.0   \n",
       "865  AAEL000249        RB            320      1.0     112.0        112.0   \n",
       "\n",
       "     overlap_group_ignoring_gene  \n",
       "864                          316  \n",
       "865                          316  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 1351.0\n",
      "overlap len: 112.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10934</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>241883099.0</td>\n",
       "      <td>241883660.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL002933-PA;Parent=AAEL002933-RA;protein_...</td>\n",
       "      <td>AAEL002933-RA</td>\n",
       "      <td>AAEL002933</td>\n",
       "      <td>RA,RB</td>\n",
       "      <td>4015</td>\n",
       "      <td>2.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10935</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>241883162.0</td>\n",
       "      <td>241883660.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL019831-PA;Parent=AAEL019831-RA;protein_...</td>\n",
       "      <td>AAEL019831-RA</td>\n",
       "      <td>AAEL019831</td>\n",
       "      <td>RB</td>\n",
       "      <td>4016</td>\n",
       "      <td>1.0</td>\n",
       "      <td>499.0</td>\n",
       "      <td>499.0</td>\n",
       "      <td>4007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "10934     1  VectorBase  CDS  241883099.0  241883660.0     .      -     0   \n",
       "10935     1  VectorBase  CDS  241883162.0  241883660.0     .      +     0   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "10934  ID=AAEL002933-PA;Parent=AAEL002933-RA;protein_...  AAEL002933-RA   \n",
       "10935  ID=AAEL019831-PA;Parent=AAEL019831-RA;protein_...  AAEL019831-RA   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "10934  AAEL002933     RA,RB           4015      2.0     562.0          6.0   \n",
       "10935  AAEL019831        RB           4016      1.0     499.0        499.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "10934                         4007  \n",
       "10935                         4007  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 562.0\n",
      "overlap len: 499.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12302</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276065573.0</td>\n",
       "      <td>276066211.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL006372-PA;Parent=AAEL006372-RA;protein_...</td>\n",
       "      <td>AAEL006372-RA</td>\n",
       "      <td>AAEL006372</td>\n",
       "      <td>RA,RC</td>\n",
       "      <td>4674</td>\n",
       "      <td>2.0</td>\n",
       "      <td>639.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>4665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12303</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276065755.0</td>\n",
       "      <td>276066211.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=AAEL006384-PA;Parent=AAEL006384-RA;protein_...</td>\n",
       "      <td>AAEL006384-RA</td>\n",
       "      <td>AAEL006384</td>\n",
       "      <td>RC</td>\n",
       "      <td>4675</td>\n",
       "      <td>1.0</td>\n",
       "      <td>457.0</td>\n",
       "      <td>457.0</td>\n",
       "      <td>4665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "12302     1  VectorBase  CDS  276065573.0  276066211.0     .      -     0   \n",
       "12303     1  VectorBase  CDS  276065755.0  276066211.0     .      -     2   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "12302  ID=AAEL006372-PA;Parent=AAEL006372-RA;protein_...  AAEL006372-RA   \n",
       "12303  ID=AAEL006384-PA;Parent=AAEL006384-RA;protein_...  AAEL006384-RA   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "12302  AAEL006372     RA,RC           4674      2.0     639.0        152.0   \n",
       "12303  AAEL006384        RC           4675      1.0     457.0        457.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "12302                         4665  \n",
       "12303                         4665  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 639.0\n",
      "overlap len: 457.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12306</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276070568.0</td>\n",
       "      <td>276070579.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL006378-PA;Parent=AAEL006378-RA;protein_...</td>\n",
       "      <td>AAEL006378-RA</td>\n",
       "      <td>AAEL006378</td>\n",
       "      <td>RA</td>\n",
       "      <td>4678</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>4668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12307</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276070568.0</td>\n",
       "      <td>276070579.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL006368-PA;Parent=AAEL006368-RA;protein_...</td>\n",
       "      <td>AAEL006368-RA</td>\n",
       "      <td>AAEL006368</td>\n",
       "      <td>RB</td>\n",
       "      <td>4679</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>4668</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "12306     1  VectorBase  CDS  276070568.0  276070579.0     .      -     0   \n",
       "12307     1  VectorBase  CDS  276070568.0  276070579.0     .      +     0   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "12306  ID=AAEL006378-PA;Parent=AAEL006378-RA;protein_...  AAEL006378-RA   \n",
       "12307  ID=AAEL006368-PA;Parent=AAEL006368-RA;protein_...  AAEL006368-RA   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "12306  AAEL006378        RA           4678      1.0      12.0         12.0   \n",
       "12307  AAEL006368        RB           4679      1.0      12.0         12.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "12306                         4668  \n",
       "12307                         4668  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 12.0\n",
      "overlap len: 12.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12312</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276083774.0</td>\n",
       "      <td>276083785.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL006376-PA;Parent=AAEL006376-RA;protein_...</td>\n",
       "      <td>AAEL006376-RA</td>\n",
       "      <td>AAEL006376</td>\n",
       "      <td>RA</td>\n",
       "      <td>4684</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>4672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12313</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276083774.0</td>\n",
       "      <td>276083785.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL006382-PB;Parent=AAEL006382-RB;protein_...</td>\n",
       "      <td>AAEL006382-RB</td>\n",
       "      <td>AAEL006382</td>\n",
       "      <td>RB</td>\n",
       "      <td>4685</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>4672</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "12312     1  VectorBase  CDS  276083774.0  276083785.0     .      -     0   \n",
       "12313     1  VectorBase  CDS  276083774.0  276083785.0     .      -     0   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "12312  ID=AAEL006376-PA;Parent=AAEL006376-RA;protein_...  AAEL006376-RA   \n",
       "12313  ID=AAEL006382-PB;Parent=AAEL006382-RB;protein_...  AAEL006382-RB   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "12312  AAEL006376        RA           4684      1.0      12.0         12.0   \n",
       "12313  AAEL006382        RB           4685      1.0      12.0         12.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "12312                         4672  \n",
       "12313                         4672  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 12.0\n",
      "overlap len: 12.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12316</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276091615.0</td>\n",
       "      <td>276091626.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL006382-PA;Parent=AAEL006382-RA;protein_...</td>\n",
       "      <td>AAEL006382-RA</td>\n",
       "      <td>AAEL006382</td>\n",
       "      <td>RA</td>\n",
       "      <td>4687</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>4674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12318</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>276091615.0</td>\n",
       "      <td>276091626.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL006383-PA;Parent=AAEL006383-RA;protein_...</td>\n",
       "      <td>AAEL006383-RA</td>\n",
       "      <td>AAEL006383</td>\n",
       "      <td>RB</td>\n",
       "      <td>4688</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>4674</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "12316     1  VectorBase  CDS  276091615.0  276091626.0     .      -     0   \n",
       "12318     1  VectorBase  CDS  276091615.0  276091626.0     .      +     0   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "12316  ID=AAEL006382-PA;Parent=AAEL006382-RA;protein_...  AAEL006382-RA   \n",
       "12318  ID=AAEL006383-PA;Parent=AAEL006383-RA;protein_...  AAEL006383-RA   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "12316  AAEL006382        RA           4687      1.0      12.0         12.0   \n",
       "12318  AAEL006383        RB           4688      1.0      12.0         12.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "12316                         4674  \n",
       "12318                         4674  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 12.0\n",
      "overlap len: 12.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22867</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>157122328.0</td>\n",
       "      <td>157122864.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL025252-PA;Parent=AAEL025252-RA;protein_...</td>\n",
       "      <td>AAEL025252-RA</td>\n",
       "      <td>AAEL025252</td>\n",
       "      <td>RA,RB</td>\n",
       "      <td>9234</td>\n",
       "      <td>2.0</td>\n",
       "      <td>537.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>9215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22868</th>\n",
       "      <td>1</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>157122760.0</td>\n",
       "      <td>157122864.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL012703-PA;Parent=AAEL012703-RA;protein_...</td>\n",
       "      <td>AAEL012703-RA</td>\n",
       "      <td>AAEL012703</td>\n",
       "      <td>RB</td>\n",
       "      <td>9235</td>\n",
       "      <td>1.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>9215</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "22867     1  VectorBase  CDS  157122328.0  157122864.0     .      -     0   \n",
       "22868     1  VectorBase  CDS  157122760.0  157122864.0     .      +     0   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "22867  ID=AAEL025252-PA;Parent=AAEL025252-RA;protein_...  AAEL025252-RA   \n",
       "22868  ID=AAEL012703-PA;Parent=AAEL012703-RA;protein_...  AAEL012703-RA   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "22867  AAEL025252     RA,RB           9234      2.0     537.0        362.0   \n",
       "22868  AAEL012703        RB           9235      1.0     105.0        105.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "22867                         9215  \n",
       "22868                         9215  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 537.0\n",
      "overlap len: 105.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49479</th>\n",
       "      <td>2</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>183898683.0</td>\n",
       "      <td>183900461.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL022351-PA;Parent=AAEL022351-RA;protein_...</td>\n",
       "      <td>AAEL022351-RA</td>\n",
       "      <td>AAEL022351</td>\n",
       "      <td>RB,RD</td>\n",
       "      <td>18239</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1779.0</td>\n",
       "      <td>1328.0</td>\n",
       "      <td>18208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49480</th>\n",
       "      <td>2</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>183900098.0</td>\n",
       "      <td>183900461.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL025226-PA;Parent=AAEL025226-RA;protein_...</td>\n",
       "      <td>AAEL025226-RA</td>\n",
       "      <td>AAEL025226</td>\n",
       "      <td>RD</td>\n",
       "      <td>18240</td>\n",
       "      <td>1.0</td>\n",
       "      <td>364.0</td>\n",
       "      <td>364.0</td>\n",
       "      <td>18208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "49479     2  VectorBase  CDS  183898683.0  183900461.0     .      +     0   \n",
       "49480     2  VectorBase  CDS  183900098.0  183900461.0     .      +     0   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "49479  ID=AAEL022351-PA;Parent=AAEL022351-RA;protein_...  AAEL022351-RA   \n",
       "49480  ID=AAEL025226-PA;Parent=AAEL025226-RA;protein_...  AAEL025226-RA   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "49479  AAEL022351     RB,RD          18239      2.0    1779.0       1328.0   \n",
       "49480  AAEL025226        RD          18240      1.0     364.0        364.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "49479                        18208  \n",
       "49480                        18208  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 1779.0\n",
      "overlap len: 364.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>splice_id</th>\n",
       "      <th>overlap_group</th>\n",
       "      <th>CDS cnt</th>\n",
       "      <th>full_len</th>\n",
       "      <th>overlap_len</th>\n",
       "      <th>overlap_group_ignoring_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56717</th>\n",
       "      <td>2</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>312826244.0</td>\n",
       "      <td>312826682.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL003602-PA;Parent=AAEL003602-RA;protein_...</td>\n",
       "      <td>AAEL003602-RA</td>\n",
       "      <td>AAEL003602</td>\n",
       "      <td>RA</td>\n",
       "      <td>21115</td>\n",
       "      <td>1.0</td>\n",
       "      <td>439.0</td>\n",
       "      <td>439.0</td>\n",
       "      <td>21081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56718</th>\n",
       "      <td>2</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>312826676.0</td>\n",
       "      <td>312827047.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=AAEL003590-PA;Parent=AAEL003590-RA;protein_...</td>\n",
       "      <td>AAEL003590-RA</td>\n",
       "      <td>AAEL003590</td>\n",
       "      <td>RA</td>\n",
       "      <td>21116</td>\n",
       "      <td>1.0</td>\n",
       "      <td>372.0</td>\n",
       "      <td>372.0</td>\n",
       "      <td>21081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid      source type        start          end score strand phase  \\\n",
       "56717     2  VectorBase  CDS  312826244.0  312826682.0     .      +     0   \n",
       "56718     2  VectorBase  CDS  312826676.0  312827047.0     .      -     0   \n",
       "\n",
       "                                              attributes  transcript_id  \\\n",
       "56717  ID=AAEL003602-PA;Parent=AAEL003602-RA;protein_...  AAEL003602-RA   \n",
       "56718  ID=AAEL003590-PA;Parent=AAEL003590-RA;protein_...  AAEL003590-RA   \n",
       "\n",
       "             gene splice_id  overlap_group  CDS cnt  full_len  overlap_len  \\\n",
       "56717  AAEL003602        RA          21115      1.0     439.0        439.0   \n",
       "56718  AAEL003590        RA          21116      1.0     372.0        372.0   \n",
       "\n",
       "       overlap_group_ignoring_gene  \n",
       "56717                        21081  \n",
       "56718                        21081  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full len: 804.0\n",
      "overlap len: 7.0\n"
     ]
    }
   ],
   "source": [
    "# @TCC NOTE:  (JUST INFORMATIVE)\n",
    "# The CDSs which overlap involving differen genes might be a problem if a chopchop target hits that overlap.\n",
    "# We will check for that below by *ensuring each chopchop target hits only one gene*\n",
    "\n",
    "# investigating overlaps between genes\n",
    "overlap = ((cds_grps['start']<=cds_grps['end'].shift(1)) & (cds_grps['seqid'].shift(1)==cds_grps['seqid']))\n",
    "cds_grps['overlap_group_ignoring_gene'] = (~overlap).cumsum()\n",
    "\n",
    "gb = cds_grps.groupby('overlap_group_ignoring_gene', dropna=False)\n",
    "\n",
    "print(\"number of overlaps involving diferent genes:\", (gb['gene'].nunique() > 1).sum())\n",
    "print(\"number of CDSs with overlaps involving different genes:\", overlap.sum())\n",
    "results_dict['overlapping CDSs with different gene ids'] = overlap.sum()\n",
    "\n",
    "# print out all of the overlap groups with multiple genes\n",
    "for i in gb.first()[gb['gene'].nunique() > 1].index.values:\n",
    "    display(gb.get_group(i))\n",
    "    print(\"full len:\", gb.get_group(i)['end'].max()-gb.get_group(i)['start'].min()+1)\n",
    "    print(\"overlap len:\", gb.get_group(i)['end'].min()-gb.get_group(i)['start'].max()+1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and filter the chopchop targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cas9 AaegL5_Cas9_all_transcripts_run\n",
      "running in 'AaegL5_Cas9_all_transcripts_run':\n",
      "/usr/bin/awk -F '\\t' 'BEGIN{OFS=FS}{if(FNR==1){if(NR==FNR){print $0, \"transcript_id\"}}else{print $0, FILENAME}}' AAEL*\n",
      "10116163 raw targets loaded\n",
      "3861017 different targets (sequence, location, gene_id)\n",
      "Cpf1 AaegL5_Cpf1_all_transcripts_run\n",
      "running in 'AaegL5_Cpf1_all_transcripts_run':\n",
      "/usr/bin/awk -F '\\t' 'BEGIN{OFS=FS}{if(FNR==1){if(NR==FNR){print $0, \"transcript_id\"}}else{print $0, FILENAME}}' AAEL*\n",
      "5725237 raw targets loaded\n",
      "2243878 different targets (sequence, location, gene_id)\n"
     ]
    }
   ],
   "source": [
    "raw_targets = OrderedDict()\n",
    "for mode, chopchop_out_dir in CHOPCHOP_OUT_DIR.items():\n",
    "    print(mode, chopchop_out_dir)\n",
    "\n",
    "    ### use awk to append the transcript_ids and merge all the chopchop outputs\n",
    "    # stream than into pandas.read_csv\n",
    "    cmd = (AWK_EXEC+\" -F \"+r\"'\\t' \"\n",
    "    \"'BEGIN{OFS=FS}{if(FNR==1){if(NR==FNR){print $0, \"+'\"transcript_id\"'+\"}}else{print $0, FILENAME}}' \"+\n",
    "    \"{}*\".format(GENE_PREFIX))\n",
    "\n",
    "    print(f\"running in '{chopchop_out_dir}':\\n{cmd}\")\n",
    "    with subprocess.Popen(cmd, cwd=chopchop_out_dir, shell=True, stdout=subprocess.PIPE) as p:\n",
    "        d = pd.read_csv(p.stdout, sep='\\t', index_col=False)\n",
    "\n",
    "    print(f\"{d.shape[0]} raw targets loaded\")\n",
    "\n",
    "    # add a gene_id column\n",
    "    d['gene_id'] = d['transcript_id'].apply(lambda x: x.rsplit(sep='-', maxsplit=1)[0])\n",
    "\n",
    "    ### drop duplicate targets\n",
    "    # chopchop duplicates some targets for unknown reasons... just exclude these\n",
    "    d = d.drop_duplicates(subset=['Target sequence', 'Genomic location','gene_id'], keep='first').sort_values('gene_id')\n",
    "    print(f\"{d.shape[0]} different targets (sequence, location, gene_id)\")\n",
    "    \n",
    "    raw_targets[mode] = d.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### Cas9\n",
      "3665000 targets at unique locations\n",
      "22261 targets have non-unique locations\n",
      "11121 locations hit by more than one target\n",
      "3665000 (99.40%) targets have unique locations\n",
      "858176 (23.42%) targets pass offtarget hits (mismatch) filters\n",
      "3016380 (82.30%) targets pass GC filters\n",
      "1946078 (53.10%) targets pass Efficiency filters\n",
      "#### Cpf1\n",
      "2126127 targets at unique locations\n",
      "22833 targets have non-unique locations\n",
      "11404 locations hit by more than one target\n",
      "2126127 (98.94%) targets have unique locations\n",
      "1695992 (79.77%) targets pass offtarget hits (mismatch) filters\n",
      "1249707 (58.78%) targets pass GC filters\n",
      "716463 (33.70%) targets pass Efficiency filters\n"
     ]
    }
   ],
   "source": [
    "targets = OrderedDict()\n",
    "for mode in raw_targets.keys():\n",
    "    print(\"####\", mode)\n",
    "\n",
    "    d = raw_targets[mode].copy()\n",
    "\n",
    "    # create chrom and pos columns from Genomic location\n",
    "    d[['chrom','pos']] = d['Genomic location'].str.split(':',n=1,expand=True)\n",
    "    d['pos'] = d['pos'].astype(int)\n",
    "    d.sort_values(['chrom', 'pos'], inplace=True)\n",
    "\n",
    "    # go ahead and just drop targets not hitting chroms of interest\n",
    "    d = d[d['chrom'].isin(CHROMS)].copy()\n",
    "    \n",
    "    # add the locations of the cut-points to the table of targets\n",
    "    d['cutpos'] = None\n",
    "    d['cutneg'] = None\n",
    "    d.loc[d['Strand']=='+','cutpos'] = d['pos'] + CUT_OFFSET_gRNApos_CDSpos[mode]\n",
    "    d.loc[d['Strand']=='+','cutneg'] = d['pos'] + CUT_OFFSET_gRNApos_CDSneg[mode]\n",
    "    d.loc[d['Strand']=='-','cutpos'] = d['pos'] + CUT_OFFSET_gRNAneg_CDSpos[mode]\n",
    "    d.loc[d['Strand']=='-','cutneg'] = d['pos'] + CUT_OFFSET_gRNAneg_CDSneg[mode]\n",
    "    assert d['cutpos'].isna().any() == False, \"Error with determining positions\"\n",
    "    assert d['cutneg'].isna().any() == False, \"Error with determining positions\"\n",
    "        \n",
    "\n",
    "    ### Some targets can hit multiple overlapping genes\n",
    "    # Just flag these for now...\n",
    "    d['flag unique location'] = ~d.duplicated(subset=['Target sequence', 'Genomic location'], keep=False)\n",
    "    print(f\"{d['flag unique location'].sum()} targets at unique locations\")\n",
    "    print(f\"{(d['flag unique location']==False).sum()} targets have non-unique locations\")\n",
    "    results_dict[f'{mode} targets with unique location'] = d['flag unique location'].sum()\n",
    "    print(f\"{len(d[~d['flag unique location']]['Genomic location'].unique())} locations hit by more than one target\")\n",
    "    print(f\"{d['flag unique location'].sum()} ({100*d['flag unique location'].sum()/d.shape[0]:.2f}%) targets have unique locations\")\n",
    "\n",
    "    ### Each target needs to have a unique genomic location for the analysis to work\n",
    "    # ensure there are no duplicate locations with the same gene_id (possibly from having variant transcripts from the same gene)\n",
    "    assert d.loc[:,['Target sequence', 'Genomic location', 'gene_id']].duplicated().sum() == 0, 'Duplicate locations hitting the same gene_id... this is not handled properly so bailing'\n",
    "    # remove targets without a unique location (hitting multiple genes) since they are non-specific\n",
    "    d = d[d['flag unique location']==True].copy() \n",
    "\n",
    "    # NOTE: Below this point, targets should be unique locations hitting a single gene\n",
    "\n",
    "    ### filters\n",
    "    # MAX_OFFTARGET_HITS = [0,0,0,0] # examplr for on MM\n",
    "    # create a numeric version of the MM3 column just taking the numbers from the '>=' results\n",
    "    d['MM3_num'] = pd.to_numeric(d['MM3'], errors='coerce')\n",
    "    d.loc[d['MM3_num'].isna(), 'MM3_num'] = pd.to_numeric(d.loc[d['MM3_num'].isna(),'MM3'].str[2:], errors='coerce')\n",
    "    d['sumMM'] = d.loc[:,['MM0','MM1','MM2','MM3_num']].sum(axis=1) # Not strictly needed, but good to keep track of accounting\n",
    "\n",
    "    d['flt MM'] = True\n",
    "    for i,col in enumerate(['MM0','MM1','MM2','MM3_num']):\n",
    "        if MAX_OFFTARGET_HITS[i] is not None:\n",
    "            d['flt MM'] = (d['flt MM']) & (d[col]<=MAX_OFFTARGET_HITS[i])\n",
    "    print(f\"{d['flt MM'].sum()} ({100*d['flt MM'].sum()/d.shape[0]:.2f}%) targets pass offtarget hits (mismatch) filters\")\n",
    "\n",
    "    ### GC filter\n",
    "    d['flt GC'] = ((d[\"GC content (%)\"] >= MIN_GC) & (d[\"GC content (%)\"] <= MAX_GC))\n",
    "    print(f\"{d['flt GC'].sum()} ({100*d['flt GC'].sum()/d.shape[0]:.2f}%) targets pass GC filters\")\n",
    "\n",
    "    ### Efficiency filter\n",
    "    if mode == 'Cas9':\n",
    "        min_efficiency = MIN_EFFICIENCY\n",
    "    elif mode == 'Cpf1':\n",
    "        min_efficiency = MIN_EFFICIENCY*100\n",
    "    else:\n",
    "        raise ValueError\n",
    "    d['flt Eff'] = d[\"Efficiency\"] >= min_efficiency\n",
    "    print(f\"{d['flt Eff'].sum()} ({100*d['flt Eff'].sum()/d.shape[0]:.2f}%) targets pass Efficiency filters\")\n",
    "\n",
    "    d.sort_values(['chrom','pos'], inplace=True)\n",
    "    d.reset_index(drop=True, inplace=True)\n",
    "    d['idx'] = d.index\n",
    "\n",
    "    results_dict[f'{mode} targets pass MM flt'] = d['flt MM'].sum()\n",
    "    results_dict[f'{mode} targets pass GC flt'] = d['flt GC'].sum()\n",
    "    results_dict[f'{mode} targets pass Eff flt'] = d['flt Eff'].sum()\n",
    "    \n",
    "    targets[mode] = d.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('total genes in gtf', 14626),\n",
       "             ('number of protein colding genes', 13601),\n",
       "             ('total coding length [bp]', 23255031),\n",
       "             ('overlapping CDSs with different gene ids', 37),\n",
       "             ('Cas9 targets with unique location', 3665000),\n",
       "             ('Cas9 targets pass MM flt', 858176),\n",
       "             ('Cas9 targets pass GC flt', 3016380),\n",
       "             ('Cas9 targets pass Eff flt', 1946078),\n",
       "             ('Cpf1 targets with unique location', 2126127),\n",
       "             ('Cpf1 targets pass MM flt', 1695992),\n",
       "             ('Cpf1 targets pass GC flt', 1249707),\n",
       "             ('Cpf1 targets pass Eff flt', 716463)])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rank</th>\n",
       "      <th>Target sequence</th>\n",
       "      <th>Genomic location</th>\n",
       "      <th>Strand</th>\n",
       "      <th>GC content (%)</th>\n",
       "      <th>Self-complementarity</th>\n",
       "      <th>MM0</th>\n",
       "      <th>MM1</th>\n",
       "      <th>MM2</th>\n",
       "      <th>MM3</th>\n",
       "      <th>Efficiency</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>cutpos</th>\n",
       "      <th>cutneg</th>\n",
       "      <th>flag unique location</th>\n",
       "      <th>MM3_num</th>\n",
       "      <th>sumMM</th>\n",
       "      <th>flt MM</th>\n",
       "      <th>flt GC</th>\n",
       "      <th>flt Eff</th>\n",
       "      <th>idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8</td>\n",
       "      <td>AATATACACACGTCGTTTCGAGG</td>\n",
       "      <td>1:31541</td>\n",
       "      <td>-</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.50</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>1</td>\n",
       "      <td>31541</td>\n",
       "      <td>31546</td>\n",
       "      <td>31546</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>CGACAGCTGCCTTTGATCGGAGG</td>\n",
       "      <td>1:31726</td>\n",
       "      <td>+</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.69</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>1</td>\n",
       "      <td>31726</td>\n",
       "      <td>31743</td>\n",
       "      <td>31743</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>5</td>\n",
       "      <td>AGCTCCTGTCCTTTTACTGGAGG</td>\n",
       "      <td>1:31874</td>\n",
       "      <td>-</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.61</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>1</td>\n",
       "      <td>31874</td>\n",
       "      <td>31879</td>\n",
       "      <td>31879</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>7</td>\n",
       "      <td>TTTATGGAGAAGAGTGCACAGGG</td>\n",
       "      <td>1:32139</td>\n",
       "      <td>-</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.53</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>1</td>\n",
       "      <td>32139</td>\n",
       "      <td>32144</td>\n",
       "      <td>32144</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>6</td>\n",
       "      <td>CAGGAATGTCGAACAAAGGTAGG</td>\n",
       "      <td>1:32387</td>\n",
       "      <td>+</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>AAEL012102</td>\n",
       "      <td>1</td>\n",
       "      <td>32387</td>\n",
       "      <td>32404</td>\n",
       "      <td>32404</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3664908</th>\n",
       "      <td>20</td>\n",
       "      <td>GCGTCCTGTTGCTCCTCAGATGG</td>\n",
       "      <td>3:409696795</td>\n",
       "      <td>+</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.55</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>3</td>\n",
       "      <td>409696795</td>\n",
       "      <td>409696812</td>\n",
       "      <td>409696812</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3664908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3664942</th>\n",
       "      <td>10</td>\n",
       "      <td>GGAGCATAGCGAATCGCTGAAGG</td>\n",
       "      <td>3:409697160</td>\n",
       "      <td>+</td>\n",
       "      <td>55</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.61</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>3</td>\n",
       "      <td>409697160</td>\n",
       "      <td>409697177</td>\n",
       "      <td>409697177</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3664942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3664943</th>\n",
       "      <td>19</td>\n",
       "      <td>TAGCGAATCGCTGAAGGCTAAGG</td>\n",
       "      <td>3:409697166</td>\n",
       "      <td>+</td>\n",
       "      <td>50</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.58</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>3</td>\n",
       "      <td>409697166</td>\n",
       "      <td>409697183</td>\n",
       "      <td>409697183</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3664943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3664944</th>\n",
       "      <td>3</td>\n",
       "      <td>ATGAGCGTGCAGTACGGCAACGG</td>\n",
       "      <td>3:409697198</td>\n",
       "      <td>-</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.67</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>3</td>\n",
       "      <td>409697198</td>\n",
       "      <td>409697203</td>\n",
       "      <td>409697203</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3664944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3664974</th>\n",
       "      <td>25</td>\n",
       "      <td>TCAGCTAGCCTAATCAAACATGG</td>\n",
       "      <td>3:409698321</td>\n",
       "      <td>+</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.51</td>\n",
       "      <td>AAEL010318-RA</td>\n",
       "      <td>AAEL010318</td>\n",
       "      <td>3</td>\n",
       "      <td>409698321</td>\n",
       "      <td>409698338</td>\n",
       "      <td>409698338</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3664974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>477435 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Rank          Target sequence Genomic location Strand  \\\n",
       "6           8  AATATACACACGTCGTTTCGAGG          1:31541      -   \n",
       "11          2  CGACAGCTGCCTTTGATCGGAGG          1:31726      +   \n",
       "20          5  AGCTCCTGTCCTTTTACTGGAGG          1:31874      -   \n",
       "34          7  TTTATGGAGAAGAGTGCACAGGG          1:32139      -   \n",
       "51          6  CAGGAATGTCGAACAAAGGTAGG          1:32387      +   \n",
       "...       ...                      ...              ...    ...   \n",
       "3664908    20  GCGTCCTGTTGCTCCTCAGATGG      3:409696795      +   \n",
       "3664942    10  GGAGCATAGCGAATCGCTGAAGG      3:409697160      +   \n",
       "3664943    19  TAGCGAATCGCTGAAGGCTAAGG      3:409697166      +   \n",
       "3664944     3  ATGAGCGTGCAGTACGGCAACGG      3:409697198      -   \n",
       "3664974    25  TCAGCTAGCCTAATCAAACATGG      3:409698321      +   \n",
       "\n",
       "         GC content (%)  Self-complementarity  MM0  MM1  MM2 MM3  Efficiency  \\\n",
       "6                    40                     0    0    0    0   0        0.50   \n",
       "11                   60                     0    0    0    0   0        0.69   \n",
       "20                   50                     0    0    0    0   0        0.61   \n",
       "34                   40                     0    0    0    0   0        0.53   \n",
       "51                   45                     0    0    0    0   0        0.59   \n",
       "...                 ...                   ...  ...  ...  ...  ..         ...   \n",
       "3664908              60                     0    0    0    0   0        0.55   \n",
       "3664942              55                     2    0    0    0   0        0.61   \n",
       "3664943              50                     3    0    0    0   0        0.58   \n",
       "3664944              55                     0    0    0    0   0        0.67   \n",
       "3664974              40                     0    0    0    0   0        0.51   \n",
       "\n",
       "         transcript_id     gene_id chrom        pos     cutpos     cutneg  \\\n",
       "6        AAEL012102-RA  AAEL012102     1      31541      31546      31546   \n",
       "11       AAEL012102-RA  AAEL012102     1      31726      31743      31743   \n",
       "20       AAEL012102-RA  AAEL012102     1      31874      31879      31879   \n",
       "34       AAEL012102-RA  AAEL012102     1      32139      32144      32144   \n",
       "51       AAEL012102-RA  AAEL012102     1      32387      32404      32404   \n",
       "...                ...         ...   ...        ...        ...        ...   \n",
       "3664908  AAEL010318-RA  AAEL010318     3  409696795  409696812  409696812   \n",
       "3664942  AAEL010318-RA  AAEL010318     3  409697160  409697177  409697177   \n",
       "3664943  AAEL010318-RA  AAEL010318     3  409697166  409697183  409697183   \n",
       "3664944  AAEL010318-RA  AAEL010318     3  409697198  409697203  409697203   \n",
       "3664974  AAEL010318-RA  AAEL010318     3  409698321  409698338  409698338   \n",
       "\n",
       "         flag unique location  MM3_num  sumMM  flt MM  flt GC  flt Eff  \\\n",
       "6                        True      0.0    0.0    True    True     True   \n",
       "11                       True      0.0    0.0    True    True     True   \n",
       "20                       True      0.0    0.0    True    True     True   \n",
       "34                       True      0.0    0.0    True    True     True   \n",
       "51                       True      0.0    0.0    True    True     True   \n",
       "...                       ...      ...    ...     ...     ...      ...   \n",
       "3664908                  True      0.0    0.0    True    True     True   \n",
       "3664942                  True      0.0    0.0    True    True     True   \n",
       "3664943                  True      0.0    0.0    True    True     True   \n",
       "3664944                  True      0.0    0.0    True    True     True   \n",
       "3664974                  True      0.0    0.0    True    True     True   \n",
       "\n",
       "             idx  \n",
       "6              6  \n",
       "11            11  \n",
       "20            20  \n",
       "34            34  \n",
       "51            51  \n",
       "...          ...  \n",
       "3664908  3664908  \n",
       "3664942  3664942  \n",
       "3664943  3664943  \n",
       "3664944  3664944  \n",
       "3664974  3664974  \n",
       "\n",
       "[477435 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "d = targets['Cas9']\n",
    "display_all_cols(d[(d['flt MM']==True) & (d['flt GC']==True) & (d['flt Eff']==True)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### Cas9\n",
      "1 +\n",
      "all looks sane\n",
      "0 of 7374 : 0.00%\n",
      "1000 of 7374 : 13.56%\n",
      "2000 of 7374 : 27.12%\n",
      "3000 of 7374 : 40.68%\n",
      "4000 of 7374 : 54.24%\n",
      "5000 of 7374 : 67.81%\n",
      "6000 of 7374 : 81.37%\n",
      "7000 of 7374 : 94.93%\n",
      "1 -\n",
      "all looks sane\n",
      "0 of 7614 : 0.00%\n",
      "1000 of 7614 : 13.13%\n",
      "2000 of 7614 : 26.27%\n",
      "3000 of 7614 : 39.40%\n",
      "4000 of 7614 : 52.53%\n",
      "5000 of 7614 : 65.67%\n",
      "6000 of 7614 : 78.80%\n",
      "7000 of 7614 : 91.94%\n",
      "3176 gene ids with a target in their CDSs\n",
      "2 +\n",
      "all looks sane\n",
      "0 of 13332 : 0.00%\n",
      "1000 of 13332 : 7.50%\n",
      "2000 of 13332 : 15.00%\n",
      "3000 of 13332 : 22.50%\n",
      "4000 of 13332 : 30.00%\n",
      "5000 of 13332 : 37.50%\n",
      "6000 of 13332 : 45.00%\n",
      "7000 of 13332 : 52.51%\n",
      "8000 of 13332 : 60.01%\n",
      "9000 of 13332 : 67.51%\n",
      "10000 of 13332 : 75.01%\n",
      "11000 of 13332 : 82.51%\n",
      "12000 of 13332 : 90.01%\n",
      "13000 of 13332 : 97.51%\n",
      "2 -\n",
      "all looks sane\n",
      "0 of 13670 : 0.00%\n",
      "1000 of 13670 : 7.32%\n",
      "2000 of 13670 : 14.63%\n",
      "3000 of 13670 : 21.95%\n",
      "4000 of 13670 : 29.26%\n",
      "5000 of 13670 : 36.58%\n",
      "6000 of 13670 : 43.89%\n",
      "7000 of 13670 : 51.21%\n",
      "8000 of 13670 : 58.52%\n",
      "9000 of 13670 : 65.84%\n",
      "10000 of 13670 : 73.15%\n",
      "11000 of 13670 : 80.47%\n",
      "12000 of 13670 : 87.78%\n",
      "13000 of 13670 : 95.10%\n",
      "8611 gene ids with a target in their CDSs\n",
      "3 +\n",
      "all looks sane\n",
      "0 of 11618 : 0.00%\n",
      "1000 of 11618 : 8.61%\n",
      "2000 of 11618 : 17.21%\n",
      "3000 of 11618 : 25.82%\n",
      "4000 of 11618 : 34.43%\n",
      "5000 of 11618 : 43.04%\n",
      "6000 of 11618 : 51.64%\n",
      "7000 of 11618 : 60.25%\n",
      "8000 of 11618 : 68.86%\n",
      "9000 of 11618 : 77.47%\n",
      "10000 of 11618 : 86.07%\n",
      "11000 of 11618 : 94.68%\n",
      "3 -\n",
      "all looks sane\n",
      "0 of 11320 : 0.00%\n",
      "1000 of 11320 : 8.83%\n",
      "2000 of 11320 : 17.67%\n",
      "3000 of 11320 : 26.50%\n",
      "4000 of 11320 : 35.34%\n",
      "5000 of 11320 : 44.17%\n",
      "6000 of 11320 : 53.00%\n",
      "7000 of 11320 : 61.84%\n",
      "8000 of 11320 : 70.67%\n",
      "9000 of 11320 : 79.51%\n",
      "10000 of 11320 : 88.34%\n",
      "11000 of 11320 : 97.17%\n",
      "13543 gene ids with a target in their CDSs\n",
      "#### Cpf1\n",
      "1 +\n",
      "all looks sane\n",
      "0 of 7374 : 0.00%\n",
      "1000 of 7374 : 13.56%\n",
      "2000 of 7374 : 27.12%\n",
      "3000 of 7374 : 40.68%\n",
      "4000 of 7374 : 54.24%\n",
      "5000 of 7374 : 67.81%\n",
      "6000 of 7374 : 81.37%\n",
      "7000 of 7374 : 94.93%\n",
      "1 -\n",
      "all looks sane\n",
      "0 of 7614 : 0.00%\n",
      "1000 of 7614 : 13.13%\n",
      "2000 of 7614 : 26.27%\n",
      "3000 of 7614 : 39.40%\n",
      "4000 of 7614 : 52.53%\n",
      "5000 of 7614 : 65.67%\n",
      "6000 of 7614 : 78.80%\n",
      "7000 of 7614 : 91.94%\n",
      "3173 gene ids with a target in their CDSs\n",
      "2 +\n",
      "all looks sane\n",
      "0 of 13332 : 0.00%\n",
      "1000 of 13332 : 7.50%\n",
      "2000 of 13332 : 15.00%\n",
      "3000 of 13332 : 22.50%\n",
      "4000 of 13332 : 30.00%\n",
      "5000 of 13332 : 37.50%\n",
      "6000 of 13332 : 45.00%\n",
      "7000 of 13332 : 52.51%\n",
      "8000 of 13332 : 60.01%\n",
      "9000 of 13332 : 67.51%\n",
      "10000 of 13332 : 75.01%\n",
      "11000 of 13332 : 82.51%\n",
      "12000 of 13332 : 90.01%\n",
      "13000 of 13332 : 97.51%\n",
      "2 -\n",
      "all looks sane\n",
      "0 of 13670 : 0.00%\n",
      "1000 of 13670 : 7.32%\n",
      "2000 of 13670 : 14.63%\n",
      "3000 of 13670 : 21.95%\n",
      "4000 of 13670 : 29.26%\n",
      "5000 of 13670 : 36.58%\n",
      "6000 of 13670 : 43.89%\n",
      "7000 of 13670 : 51.21%\n",
      "8000 of 13670 : 58.52%\n",
      "9000 of 13670 : 65.84%\n",
      "10000 of 13670 : 73.15%\n",
      "11000 of 13670 : 80.47%\n",
      "12000 of 13670 : 87.78%\n",
      "13000 of 13670 : 95.10%\n",
      "8606 gene ids with a target in their CDSs\n",
      "3 +\n",
      "all looks sane\n",
      "0 of 11618 : 0.00%\n",
      "1000 of 11618 : 8.61%\n",
      "2000 of 11618 : 17.21%\n",
      "3000 of 11618 : 25.82%\n",
      "4000 of 11618 : 34.43%\n",
      "5000 of 11618 : 43.04%\n",
      "6000 of 11618 : 51.64%\n",
      "7000 of 11618 : 60.25%\n",
      "8000 of 11618 : 68.86%\n",
      "9000 of 11618 : 77.47%\n",
      "10000 of 11618 : 86.07%\n",
      "11000 of 11618 : 94.68%\n",
      "3 -\n",
      "all looks sane\n",
      "0 of 11320 : 0.00%\n",
      "1000 of 11320 : 8.83%\n",
      "2000 of 11320 : 17.67%\n",
      "3000 of 11320 : 26.50%\n",
      "4000 of 11320 : 35.34%\n",
      "5000 of 11320 : 44.17%\n",
      "6000 of 11320 : 53.00%\n",
      "7000 of 11320 : 61.84%\n",
      "8000 of 11320 : 70.67%\n",
      "9000 of 11320 : 79.51%\n",
      "10000 of 11320 : 88.34%\n",
      "11000 of 11320 : 97.17%\n",
      "13538 gene ids with a target in their CDSs\n"
     ]
    }
   ],
   "source": [
    "# %%time\n",
    "# mapping targets to CDSs\n",
    "\n",
    "for mode in targets.keys():\n",
    "    print('####', mode)\n",
    "    \n",
    "    d = targets[mode].copy() \n",
    "\n",
    "    d['gene_id'] = None\n",
    "\n",
    "    ### Targets where cutpoint hits a CDS\n",
    "    for chrom in CHROMS:\n",
    "        \n",
    "        for cds_strand,cut_col in [('+','cutpos'), ('-','cutneg')]: \n",
    "            print(chrom, cds_strand)\n",
    "            cdstmp = cds.loc[(cds['seqid']==chrom) & (cds['strand']==cds_strand),:]       \n",
    "            tmp = d.loc[d['chrom']==chrom].sort_values(cut_col)\n",
    "            starts_idx = np.searchsorted(tmp[cut_col].values, cdstmp['start'].values, side='left')\n",
    "            ends_idx = np.searchsorted(tmp[cut_col].values, cdstmp['end'].values, side='right')\n",
    "        \n",
    "     \n",
    "            # some sanity testing\n",
    "            cutpos_col_num = tmp.columns.get_loc(cut_col)\n",
    "            for i in range(cdstmp.shape[0]):\n",
    "                # if cdstmp.iloc[i]['end'] == tmp.iloc[ends_idx[i]-1,cutpos_col_num]:\n",
    "                #     print(f\"FOO: {cdstmp.iloc[i]['end']} == {tmp.iloc[ends_idx[i]-1,cutpos_col_num]}\")\n",
    "                #     display(cdstmp.iloc[i])\n",
    "                #     display(tmp.iloc[ends_idx[i]-1,:])\n",
    "                #     break\n",
    "\n",
    "                if i>0 and cdstmp.iloc[i]['start'] < tmp.iloc[starts_idx[i]-1,cutpos_col_num]:\n",
    "                    print(\"target in cds start missed\")\n",
    "                    break\n",
    "                if cdstmp.iloc[i]['start'] > tmp.iloc[starts_idx[i],cutpos_col_num]:\n",
    "                    print(\"first target is before cds start\")\n",
    "                    break\n",
    "                if cdstmp.iloc[i]['end'] < tmp.iloc[ends_idx[i]-1,cutpos_col_num]:\n",
    "                    print(\"end target is after cds end\")\n",
    "                    break\n",
    "                if i<len(ends_idx)-1 and cdstmp.iloc[i]['end'] > tmp.iloc[ends_idx[i],cutpos_col_num]:\n",
    "                    print(\"target in cds end missed\")\n",
    "                    break\n",
    "            if i < len(ends_idx)-1:\n",
    "                print(f\"{tmp.iloc[starts_idx[i]-1,cutpos_col_num]} ({cds.iloc[i]['start']}) {tmp.iloc[starts_idx[i],cutpos_col_num]} ... {tmp.iloc[ends_idx[i]-1,cutpos_col_num]} ({cds.iloc[i]['end']}) {tmp.iloc[ends_idx[i],cutpos_col_num]}\")\n",
    "            else:\n",
    "                print(\"all looks sane\")\n",
    "\n",
    "            # Assigning gene_id to each target\n",
    "            tmp_idx_col_num = tmp.columns.get_loc('idx') # the 'idx' col maps values from tmp (slice of d) back to the index of d\n",
    "            for i in range(cdstmp.shape[0]):\n",
    "                if i%1000 == 0:\n",
    "                    print(f\"{i} of {cdstmp.shape[0]} : {i*100/cdstmp.shape[0]:2.2f}%\")\n",
    "                if starts_idx[i]<ends_idx[i]: \n",
    "                    gene_id = cdstmp.iloc[i]['gene']\n",
    "                    # print(starts_idx[i], ends_idx[i], gene_id)\n",
    "                    # display_all_cols(d.loc[tmp.iloc[starts_idx[i]:ends_idx[i],tmp_idx_col_num].values,:])\n",
    "                    assert d.loc[tmp.iloc[starts_idx[i]:ends_idx[i],tmp_idx_col_num].values,'gene_id'].isin([None, gene_id]).all(), \"problem: target appears to hit multiple genes\"\n",
    "                    d.loc[tmp.iloc[starts_idx[i]:ends_idx[i],tmp_idx_col_num].values,'gene_id'] = gene_id\n",
    "                    # display_all_cols(d.loc[tmp.iloc[starts_idx[i]:ends_idx[i],tmp_idx_col_num].values,:])\n",
    "        print(f\"{d['gene_id'].nunique()-1} gene ids with a target in their CDSs\")\n",
    "        # more sanity checking\n",
    "        assert (d.loc[:,('chrom','pos')].duplicated() == False).all(), \"All targets should be unique locations\"\n",
    "       \n",
    "        # save results\n",
    "        targets[mode] = d.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rank</th>\n",
       "      <th>Target sequence</th>\n",
       "      <th>Genomic location</th>\n",
       "      <th>Strand</th>\n",
       "      <th>GC content (%)</th>\n",
       "      <th>Self-complementarity</th>\n",
       "      <th>MM0</th>\n",
       "      <th>MM1</th>\n",
       "      <th>MM2</th>\n",
       "      <th>MM3</th>\n",
       "      <th>Efficiency</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>cutpos</th>\n",
       "      <th>cutneg</th>\n",
       "      <th>flag unique location</th>\n",
       "      <th>MM3_num</th>\n",
       "      <th>sumMM</th>\n",
       "      <th>flt MM</th>\n",
       "      <th>flt GC</th>\n",
       "      <th>flt Eff</th>\n",
       "      <th>idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>102</td>\n",
       "      <td>ATAGCAAACAAGGACCAGCCTGG</td>\n",
       "      <td>1:41748</td>\n",
       "      <td>-</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.49</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>41748</td>\n",
       "      <td>41753</td>\n",
       "      <td>41753</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>86</td>\n",
       "      <td>CTGGTCCTTGTTTGCTATTGTGG</td>\n",
       "      <td>1:41753</td>\n",
       "      <td>+</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.37</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41753</td>\n",
       "      <td>41770</td>\n",
       "      <td>41770</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>9</td>\n",
       "      <td>GTCCTTGTTTGCTATTGTGGTGG</td>\n",
       "      <td>1:41756</td>\n",
       "      <td>+</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41756</td>\n",
       "      <td>41773</td>\n",
       "      <td>41773</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>49</td>\n",
       "      <td>GTCCACCACAATAGCAAACAAGG</td>\n",
       "      <td>1:41758</td>\n",
       "      <td>-</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.50</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>41758</td>\n",
       "      <td>41763</td>\n",
       "      <td>41763</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>12</td>\n",
       "      <td>TTTGCTATTGTGGTGGACACCGG</td>\n",
       "      <td>1:41763</td>\n",
       "      <td>+</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.52</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41763</td>\n",
       "      <td>41780</td>\n",
       "      <td>41780</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>32</td>\n",
       "      <td>TAGCTACTTCAACGGTAGTCCGG</td>\n",
       "      <td>1:41782</td>\n",
       "      <td>-</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.37</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41782</td>\n",
       "      <td>41787</td>\n",
       "      <td>41787</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>139</td>\n",
       "      <td>CGATGAATTAGCTACTTCAACGG</td>\n",
       "      <td>1:41790</td>\n",
       "      <td>-</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.49</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41790</td>\n",
       "      <td>41795</td>\n",
       "      <td>41795</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>157</td>\n",
       "      <td>TCATCGAATGTTTTATCGATAGG</td>\n",
       "      <td>1:41807</td>\n",
       "      <td>+</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.47</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41807</td>\n",
       "      <td>41824</td>\n",
       "      <td>41824</td>\n",
       "      <td>True</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>70</td>\n",
       "      <td>ATTCGATCGTCATACAACGTTGG</td>\n",
       "      <td>1:41831</td>\n",
       "      <td>+</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.55</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41831</td>\n",
       "      <td>41848</td>\n",
       "      <td>41848</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>15</td>\n",
       "      <td>GCGGTACTGGCTGGAACTGCTGG</td>\n",
       "      <td>1:41870</td>\n",
       "      <td>-</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.50</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41870</td>\n",
       "      <td>41875</td>\n",
       "      <td>41875</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>82</td>\n",
       "      <td>TTCACGTTAGCGGTACTGGCTGG</td>\n",
       "      <td>1:41879</td>\n",
       "      <td>-</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.43</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41879</td>\n",
       "      <td>41884</td>\n",
       "      <td>41884</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>108</td>\n",
       "      <td>GATCTTCACGTTAGCGGTACTGG</td>\n",
       "      <td>1:41883</td>\n",
       "      <td>-</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.37</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41883</td>\n",
       "      <td>41888</td>\n",
       "      <td>41888</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>54</td>\n",
       "      <td>CTTGTGGATCTTCACGTTAGCGG</td>\n",
       "      <td>1:41889</td>\n",
       "      <td>-</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.46</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41889</td>\n",
       "      <td>41894</td>\n",
       "      <td>41894</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>122</td>\n",
       "      <td>TTTTGGATCGGTTATTCTTGTGG</td>\n",
       "      <td>1:41905</td>\n",
       "      <td>-</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.45</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41905</td>\n",
       "      <td>41910</td>\n",
       "      <td>41910</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>101</td>\n",
       "      <td>ACGTTGCAATGGTTTTGGATCGG</td>\n",
       "      <td>1:41917</td>\n",
       "      <td>-</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.49</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41917</td>\n",
       "      <td>41922</td>\n",
       "      <td>41922</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>114</td>\n",
       "      <td>GAAATACGTTGCAATGGTTTTGG</td>\n",
       "      <td>1:41922</td>\n",
       "      <td>-</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.33</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41922</td>\n",
       "      <td>41927</td>\n",
       "      <td>41927</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>142</td>\n",
       "      <td>GGATTGGAAATACGTTGCAATGG</td>\n",
       "      <td>1:41928</td>\n",
       "      <td>-</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.53</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41928</td>\n",
       "      <td>41933</td>\n",
       "      <td>41933</td>\n",
       "      <td>True</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>201</td>\n",
       "      <td>TATTGCAGGTGAAAGAGGATTGG</td>\n",
       "      <td>1:41944</td>\n",
       "      <td>-</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>&gt;=307</td>\n",
       "      <td>0.48</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41944</td>\n",
       "      <td>41949</td>\n",
       "      <td>41949</td>\n",
       "      <td>True</td>\n",
       "      <td>307.0</td>\n",
       "      <td>316.0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>69</td>\n",
       "      <td>ACTTATATTGCAGGTGAAAGAGG</td>\n",
       "      <td>1:41949</td>\n",
       "      <td>-</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.56</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>AAEL012106</td>\n",
       "      <td>1</td>\n",
       "      <td>41949</td>\n",
       "      <td>41954</td>\n",
       "      <td>41954</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>83</td>\n",
       "      <td>CACACTTCTACTTATATTGCAGG</td>\n",
       "      <td>1:41958</td>\n",
       "      <td>-</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.42</td>\n",
       "      <td>AAEL012106-RA</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>41958</td>\n",
       "      <td>41963</td>\n",
       "      <td>41963</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Rank          Target sequence Genomic location Strand  GC content (%)  \\\n",
       "153   102  ATAGCAAACAAGGACCAGCCTGG          1:41748      -              50   \n",
       "154    86  CTGGTCCTTGTTTGCTATTGTGG          1:41753      +              45   \n",
       "155     9  GTCCTTGTTTGCTATTGTGGTGG          1:41756      +              45   \n",
       "156    49  GTCCACCACAATAGCAAACAAGG          1:41758      -              45   \n",
       "157    12  TTTGCTATTGTGGTGGACACCGG          1:41763      +              45   \n",
       "158    32  TAGCTACTTCAACGGTAGTCCGG          1:41782      -              45   \n",
       "159   139  CGATGAATTAGCTACTTCAACGG          1:41790      -              35   \n",
       "160   157  TCATCGAATGTTTTATCGATAGG          1:41807      +              30   \n",
       "161    70  ATTCGATCGTCATACAACGTTGG          1:41831      +              40   \n",
       "162    15  GCGGTACTGGCTGGAACTGCTGG          1:41870      -              65   \n",
       "163    82  TTCACGTTAGCGGTACTGGCTGG          1:41879      -              55   \n",
       "164   108  GATCTTCACGTTAGCGGTACTGG          1:41883      -              50   \n",
       "165    54  CTTGTGGATCTTCACGTTAGCGG          1:41889      -              45   \n",
       "166   122  TTTTGGATCGGTTATTCTTGTGG          1:41905      -              35   \n",
       "167   101  ACGTTGCAATGGTTTTGGATCGG          1:41917      -              40   \n",
       "168   114  GAAATACGTTGCAATGGTTTTGG          1:41922      -              35   \n",
       "169   142  GGATTGGAAATACGTTGCAATGG          1:41928      -              40   \n",
       "170   201  TATTGCAGGTGAAAGAGGATTGG          1:41944      -              40   \n",
       "171    69  ACTTATATTGCAGGTGAAAGAGG          1:41949      -              35   \n",
       "172    83  CACACTTCTACTTATATTGCAGG          1:41958      -              35   \n",
       "\n",
       "     Self-complementarity  MM0  MM1  MM2    MM3  Efficiency  transcript_id  \\\n",
       "153                     1    0    0    0      3        0.49  AAEL012106-RA   \n",
       "154                     0    0    0    0      2        0.37  AAEL012106-RA   \n",
       "155                     0    0    0    0      0        0.59  AAEL012106-RA   \n",
       "156                     0    0    0    0      1        0.50  AAEL012106-RA   \n",
       "157                     0    0    0    0      0        0.52  AAEL012106-RA   \n",
       "158                     1    0    0    0      0        0.37  AAEL012106-RA   \n",
       "159                     0    0    0    3      2        0.49  AAEL012106-RA   \n",
       "160                     2    0    0    1      7        0.47  AAEL012106-RA   \n",
       "161                     0    0    0    0      2        0.55  AAEL012106-RA   \n",
       "162                     0    0    0    0      0        0.50  AAEL012106-RA   \n",
       "163                     0    0    0    0      2        0.43  AAEL012106-RA   \n",
       "164                     0    0    0    0      3        0.37  AAEL012106-RA   \n",
       "165                     0    0    0    0      1        0.46  AAEL012106-RA   \n",
       "166                     0    0    0    0      4        0.45  AAEL012106-RA   \n",
       "167                     0    0    0    0      3        0.49  AAEL012106-RA   \n",
       "168                     0    0    0    1      2        0.33  AAEL012106-RA   \n",
       "169                     0    0    0    0      7        0.53  AAEL012106-RA   \n",
       "170                     0    0    0    9  >=307        0.48  AAEL012106-RA   \n",
       "171                     0    0    0    0      2        0.56  AAEL012106-RA   \n",
       "172                     0    0    0    0      2        0.42  AAEL012106-RA   \n",
       "\n",
       "        gene_id chrom    pos cutpos cutneg  flag unique location  MM3_num  \\\n",
       "153        None     1  41748  41753  41753                  True      3.0   \n",
       "154  AAEL012106     1  41753  41770  41770                  True      2.0   \n",
       "155  AAEL012106     1  41756  41773  41773                  True      0.0   \n",
       "156        None     1  41758  41763  41763                  True      1.0   \n",
       "157  AAEL012106     1  41763  41780  41780                  True      0.0   \n",
       "158  AAEL012106     1  41782  41787  41787                  True      0.0   \n",
       "159  AAEL012106     1  41790  41795  41795                  True      2.0   \n",
       "160  AAEL012106     1  41807  41824  41824                  True      7.0   \n",
       "161  AAEL012106     1  41831  41848  41848                  True      2.0   \n",
       "162  AAEL012106     1  41870  41875  41875                  True      0.0   \n",
       "163  AAEL012106     1  41879  41884  41884                  True      2.0   \n",
       "164  AAEL012106     1  41883  41888  41888                  True      3.0   \n",
       "165  AAEL012106     1  41889  41894  41894                  True      1.0   \n",
       "166  AAEL012106     1  41905  41910  41910                  True      4.0   \n",
       "167  AAEL012106     1  41917  41922  41922                  True      3.0   \n",
       "168  AAEL012106     1  41922  41927  41927                  True      2.0   \n",
       "169  AAEL012106     1  41928  41933  41933                  True      7.0   \n",
       "170  AAEL012106     1  41944  41949  41949                  True    307.0   \n",
       "171  AAEL012106     1  41949  41954  41954                  True      2.0   \n",
       "172        None     1  41958  41963  41963                  True      2.0   \n",
       "\n",
       "     sumMM  flt MM  flt GC  flt Eff  idx  \n",
       "153    3.0   False    True    False  153  \n",
       "154    2.0   False    True    False  154  \n",
       "155    0.0    True    True     True  155  \n",
       "156    1.0   False    True     True  156  \n",
       "157    0.0    True    True     True  157  \n",
       "158    0.0    True    True    False  158  \n",
       "159    5.0   False   False    False  159  \n",
       "160    8.0   False   False    False  160  \n",
       "161    2.0   False    True     True  161  \n",
       "162    0.0    True    True     True  162  \n",
       "163    2.0   False    True    False  163  \n",
       "164    3.0   False    True    False  164  \n",
       "165    1.0   False    True    False  165  \n",
       "166    4.0   False   False    False  166  \n",
       "167    3.0   False    True    False  167  \n",
       "168    3.0   False   False    False  168  \n",
       "169    7.0   False    True     True  169  \n",
       "170  316.0   False    True    False  170  \n",
       "171    2.0   False   False     True  171  \n",
       "172    2.0   False   False    False  172  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display_all_cols(targets['Cas9'].iloc[153:173,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a flt column for if it hits a CDS\n",
    "for mode in targets.keys():\n",
    "    targets[mode]['flt hitsCDS'] = False\n",
    "    targets[mode].loc[targets[mode]['gene_id'].notna(),'flt hitsCDS'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### Cas9\n",
      "2632337 (72%) targets hit a CDS\n",
      "targets pass all fileters: 379147 (10.35%)\n",
      "#### Cpf1\n",
      "977085 (46%) targets hit a CDS\n",
      "targets pass all fileters: 265419 (12.48%)\n"
     ]
    }
   ],
   "source": [
    "# filter for targets hitting a CDS (have a gene_id) and apply previous filters to get potential_targets\n",
    "\n",
    "potential_targets = OrderedDict()\n",
    "\n",
    "for mode in targets.keys():\n",
    "    print('####', mode)\n",
    "    d = targets[mode]\n",
    "    potential_targets[mode] = d[(d['flt MM']==True) & (d['flt GC']==True) & (d['flt Eff']==True) & (d['gene_id'].notna())].copy()\n",
    "\n",
    "    tmp = d.loc[d['gene_id'].notna(),:].shape[0]    \n",
    "    results_dict[f'{mode} targets hit CDS'] = tmp\n",
    "    print(f\"{tmp} ({100*tmp/d.shape[0]:.2g}%) targets hit a CDS\")\n",
    "        \n",
    "    results_dict[f'{mode} targets pass all fileters'] = potential_targets[mode].shape[0]\n",
    "    print(f\"targets pass all fileters: {results_dict[f'{mode} targets pass all fileters']} ({100*results_dict[f'{mode} targets pass all fileters']/d.shape[0]:.2f}%)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('total genes in gtf', 14626),\n",
       "             ('number of protein colding genes', 13601),\n",
       "             ('total coding length [bp]', 23255031),\n",
       "             ('overlapping CDSs with different gene ids', 37),\n",
       "             ('Cas9 targets with unique location', 3665000),\n",
       "             ('Cas9 targets pass MM flt', 858176),\n",
       "             ('Cas9 targets pass GC flt', 3016380),\n",
       "             ('Cas9 targets pass Eff flt', 1946078),\n",
       "             ('Cpf1 targets with unique location', 2126127),\n",
       "             ('Cpf1 targets pass MM flt', 1695992),\n",
       "             ('Cpf1 targets pass GC flt', 1249707),\n",
       "             ('Cpf1 targets pass Eff flt', 716463),\n",
       "             ('Cas9 targets hit CDS', 2632337),\n",
       "             ('Cas9 targets pass all fileters', 379147),\n",
       "             ('Cpf1 targets hit CDS', 977085),\n",
       "             ('Cpf1 targets pass all fileters', 265419)])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# display_all(targets['Cpf1'].iloc[40:100,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering based on variant frequency from VCF (@TCC Modified to save all targets, not just potential)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## search vcf for target sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-07-09T19:04:43.235590Z",
     "start_time": "2019-07-09T19:04:43.000696Z"
    }
   },
   "outputs": [],
   "source": [
    "### Open the callset (zarr verison of the vcf) and get the list of CHROMS\n",
    "callset = zarr.open_group(ZVCF, mode='r')\n",
    "if CHROMS is None: # use all chroms?\n",
    "    CHROMS = list(callset.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-07-09T19:04:44.542023Z",
     "start_time": "2019-07-09T19:04:43.900620Z"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of samples 89\n"
     ]
    }
   ],
   "source": [
    "### get the (optional) subset of samples to use\n",
    "\n",
    "if SAMPLE_LIST_FN is None: # use all samples\n",
    "    meta = pd.DataFrame(list(callset.values())[0]['samples'][:])\n",
    "    meta['callset_idx'] = range(meta.shape[0])\n",
    "else:\n",
    "    # load the sample list and make meta dataframe\n",
    "    if isinstance(SAMPLE_LIST_FN, str): # assume a simple list file\n",
    "        meta = pd.read_csv(SAMPLE_LIST_FN, comment='#', header=None, index_col=0, skiprows=1)\n",
    "    else: # parse ag1000g samples.meta.txt with SAMPLE_LIST_FN really being [M|S,filename]\n",
    "        meta = pd.read_csv(SAMPLE_LIST_FN[1], delimiter='\\t', comment='#', index_col=0)\n",
    "        meta = meta[meta['m_s']==SAMPLE_LIST_FN[0]]\n",
    "    all_callset_samples = list(list(callset.values())[0]['samples'])\n",
    "    meta['callset_idx'] = [all_callset_samples.index(x) for x in meta.index]\n",
    "meta.index.name = 'sample'\n",
    "\n",
    "# Optionally subsample to check sample size effects\n",
    "if SUBSAMPLE_N:\n",
    "    meta = meta.sample(n=SUBSAMPLE_N)\n",
    "\n",
    "# ensure it is sorted by callset_idx (otherwise makes later steps terribly inefficient)\n",
    "meta = meta.sort_values('callset_idx')\n",
    "\n",
    "print(\"number of samples\",meta.shape[0])\n",
    "results_dict['number of samples'] = meta.shape[0]\n",
    "# meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-12T19:42:57.468474Z",
     "start_time": "2019-06-12T19:42:57.456453Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrom: 1\n",
      "variant filter passes: 37857469\n",
      "variants : 37857469 of 42125744 = 89.868%\n",
      "chrom: 2\n",
      "variant filter passes: 67942773\n",
      "variants : 67942773 of 74616963 = 91.055%\n",
      "chrom: 3\n",
      "variant filter passes: 56085822\n",
      "variants : 56085822 of 61792617 = 90.765%\n",
      "total variants : 161886064\n"
     ]
    }
   ],
   "source": [
    "### get variant positiosn from the callset (zvcf)\n",
    "ac_dict = OrderedDict()\n",
    "refAF_dict = OrderedDict()\n",
    "vpos_dict = OrderedDict()\n",
    "results_dict['vcf variants'] = 0\n",
    "results_dict['variants'] = 0\n",
    "\n",
    "for chrom in CHROMS:\n",
    "    print(\"chrom:\", chrom)\n",
    "    \n",
    "    ## Get the positions of variants on this chrom (possible subsetting samples)\n",
    "    pos = allel.SortedIndex(callset[chrom]['variants/POS'])\n",
    "    g = allel.GenotypeDaskArray(callset[chrom]['calldata/GT']).subset(None, meta['callset_idx'].values)\n",
    "    n_variants_in = g.shape[0]\n",
    "    \n",
    "    # ensure number of samples is same\n",
    "    assert meta.shape[0] == g.shape[1]\n",
    "    \n",
    "    # get major allele frequencies\n",
    "    ac = g.count_alleles().compute()\n",
    "    # only variant alleles matter, so only take those\n",
    "    flt_var = ac.is_variant()\n",
    "    print('variant filter passes:', flt_var.sum())\n",
    "    \n",
    "    if IGNORE_ONE_CALL_VARIANTS:\n",
    "        flt_one_call = ac[:,1:].sum(axis=1)>1\n",
    "        print('ignore-one-call variant filter passes:', flt_one_call.sum())\n",
    "        flt_var = (flt_var & flt_one_call)\n",
    "    \n",
    "    ac = ac.compress(flt_var, axis=0)\n",
    "    n_variants_var = ac.shape[0]\n",
    "\n",
    "    ac_dict[chrom] = ac\n",
    "    refAF_dict[chrom] = ac.to_frequencies(fill=1)[:,0] # ref should be allele index 0\n",
    "    vpos_dict[chrom] = pos.compress(flt_var, axis=0)[:] # `[:]` to loads into memory\n",
    "    \n",
    "    print(\"variants : {} of {} = {:.3f}%\".format(n_variants_var, n_variants_in,\n",
    "                                                 100*n_variants_var/n_variants_in))\n",
    "    results_dict['vcf variants'] += n_variants_in\n",
    "    results_dict['variants'] += n_variants_var\n",
    "        \n",
    "print('total variants :', results_dict['variants'])\n",
    "# free up memory\n",
    "del pos\n",
    "del g\n",
    "del ac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 pi: mean=0.013  std=0.012\toverall_pi=0.0153\n",
      "2 pi: mean=0.011  std=0.01\toverall_pi=0.0125\n",
      "3 pi: mean=0.011  std=0.01\toverall_pi=0.0124\n",
      "** nucleotide diversity in coding regions: 0.013172444056535816\n",
      "CPU times: user 4.27 s, sys: 4.21 ms, total: 4.27 s\n",
      "Wall time: 4.29 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "##### Nucleotide diveristiy of coding regions calcluation #####\n",
    "cds_grp_nuc_diversity = OrderedDict()\n",
    "cds_grp_length = OrderedDict()\n",
    "for chrom in CHROMS:\n",
    "\n",
    "    gb = cds[cds['seqid']==chrom].groupby('overlap_group_ignoring_gene', dropna=False)\n",
    "    starts = gb['start'].min().values\n",
    "    ends = gb['end'].max().values\n",
    "\n",
    "    cds_grp_nuc_diversity[chrom] = np.zeros(starts.shape)\n",
    "    cds_grp_length[chrom] = np.zeros(starts.shape)\n",
    "\n",
    "    for i in range(len(starts)):\n",
    "        # print(starts[i], ends[i])\n",
    "        cds_grp_length[chrom][i] = ends[i]-starts[i]+1\n",
    "        try:\n",
    "            cds_grp_nuc_diversity[chrom][i] = allel.sequence_diversity(vpos_dict[chrom], ac_dict[chrom], start=starts[i], stop=ends[i])\n",
    "        except KeyError as e:\n",
    "            assert starts[i] < ends[i] \n",
    "            # print(\"Warning: no variants in pos range\",e)\n",
    "            cds_grp_nuc_diversity[chrom][i] = 0\n",
    "    # if i>100: break\n",
    "\n",
    "\n",
    "s = 0\n",
    "n = 0\n",
    "for chrom in CHROMS:\n",
    "    print(f\"{chrom} pi: mean={np.mean(cds_grp_nuc_diversity[chrom]):.2g}  std={np.std(cds_grp_nuc_diversity[chrom]):.2g}\", end=\"\\t\")\n",
    "    print(f\"overall_pi={(cds_grp_nuc_diversity[chrom]*cds_grp_length[chrom]).sum() / cds_grp_length[chrom].sum():.3g}\")\n",
    "    s += (cds_grp_nuc_diversity[chrom]*cds_grp_length[chrom]).sum()\n",
    "    n += cds_grp_length[chrom].sum()\n",
    "print(\"** nucleotide diversity in coding regions:\", s/n)\n",
    "results_dict['nucleotide diversity in coding regions'] = s/n\n",
    "\n",
    "# @TCC: here's a one-liner to calculate total CDS pi \n",
    "# (np.concatenate(list(cds_grp_nuc_diversity.values())) * np.concatenate(list(cds_grp_length.values()))).sum() / np.concatenate(list(cds_grp_length.values())).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig, ax = mpl.pyplot.subplots(1,1, facecolor='white')\n",
    "# cnts,bins,patches = ax.hist(np.concatenate(list(cds_grp_nuc_diversity.values())), bins=500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('total genes in gtf', 14626),\n",
       "             ('number of protein colding genes', 13601),\n",
       "             ('total coding length [bp]', 23255031),\n",
       "             ('overlapping CDSs with different gene ids', 37),\n",
       "             ('Cas9 targets with unique location', 3665000),\n",
       "             ('Cas9 targets pass MM flt', 858176),\n",
       "             ('Cas9 targets pass GC flt', 3016380),\n",
       "             ('Cas9 targets pass Eff flt', 1946078),\n",
       "             ('Cpf1 targets with unique location', 2126127),\n",
       "             ('Cpf1 targets pass MM flt', 1695992),\n",
       "             ('Cpf1 targets pass GC flt', 1249707),\n",
       "             ('Cpf1 targets pass Eff flt', 716463),\n",
       "             ('Cas9 targets hit CDS', 2632337),\n",
       "             ('Cas9 targets pass all fileters', 379147),\n",
       "             ('Cpf1 targets hit CDS', 977085),\n",
       "             ('Cpf1 targets pass all fileters', 265419),\n",
       "             ('number of samples', 89),\n",
       "             ('vcf variants', 178535324),\n",
       "             ('variants', 161886064),\n",
       "             ('nucleotide diversity in coding regions', 0.013172444056535816)])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-12T19:48:20.439877Z",
     "start_time": "2019-06-12T19:48:00.158954Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cas9 23\n",
      "chrom: 1\n",
      "unique (location,gene) targets: (928704,)\n",
      "chrom: 2\n",
      "unique (location,gene) targets: (1435679,)\n",
      "chrom: 3\n",
      "unique (location,gene) targets: (1300617,)\n",
      "Cpf1 28\n",
      "chrom: 1\n",
      "unique (location,gene) targets: (483723,)\n",
      "chrom: 2\n",
      "unique (location,gene) targets: (883805,)\n",
      "chrom: 3\n",
      "unique (location,gene) targets: (758599,)\n"
     ]
    }
   ],
   "source": [
    "## Overlap between targets and variants\n",
    "\n",
    "for mode in ('Cas9', 'Cpf1'):\n",
    "    # all target sequences for a given mode are the same length\n",
    "    target_len = len(targets[mode].iloc[0]['Target sequence'])\n",
    "    print(mode, target_len)\n",
    "\n",
    "    tout = targets[mode].copy(deep=True)\n",
    "\n",
    "    tout['vpos'] = np.nan # position of each variant\n",
    "    tout['vpos'] = tout['vpos'].astype('object')\n",
    "    tout['nv'] = np.nan # number of variants\n",
    "    tout['refAF'] = np.nan # frequency of reference allele of each variant\n",
    "    tout['refAF'] = tout['refAF'].astype('object')\n",
    "    tout['p_ref'] = np.nan # probability of ref sequence for whole target\n",
    "\n",
    "    for chrom in CHROMS:\n",
    "        print(\"chrom:\", chrom)\n",
    "        vpos = vpos_dict[chrom]\n",
    "\n",
    "        # target positions (target list is filtered above)\n",
    "        tpos = targets[mode][targets[mode]['chrom']==chrom]['pos'].copy(deep=True)\n",
    "        print(\"unique (location,gene) targets:\", tpos.shape)\n",
    "        \n",
    "        # find target positions in the list of variant positions (using sorted insert positions)\n",
    "        p1 = np.searchsorted(vpos, tpos.values, side='left')\n",
    "        p2 = np.searchsorted(vpos, tpos.values+target_len, side='left')\n",
    "\n",
    "        # list of actual variant positions in each target\n",
    "        tout.loc[tpos.index,'vpos'] = np.array([vpos[_p1:_p2].values for _p1,_p2 in zip(p1,p2)], dtype=object)\n",
    "        # count of variants in each target\n",
    "        tout.loc[tpos.index,'nv'] = (p2-p1)\n",
    "        # refAF of each variant in each target\n",
    "        tout.loc[tpos.index,'refAF'] = np.array([refAF_dict[chrom][_p1:_p2] for _p1,_p2 in zip(p1,p2)], dtype=object)\n",
    "        # the probability a target locus will be entirely reference (perfect match)\n",
    "        # assumes each variant hitting that target is independent\n",
    "        tout.loc[tpos.index,'p_ref'] = tout.loc[tpos.index,'refAF'].apply(lambda x: np.prod(x, initial=1))\n",
    "\n",
    "    # make sure everything got filled in\n",
    "    assert tout[tout['nv'].isna()].empty\n",
    "    #\n",
    "    targets[mode] = tout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# free up memory by discarding the ac_dict, refAF_dict, vpos_dict  \n",
    "del ac_dict\n",
    "del refAF_dict\n",
    "del vpos_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('total genes in gtf', 14626),\n",
       "             ('number of protein colding genes', 13601),\n",
       "             ('total coding length [bp]', 23255031),\n",
       "             ('overlapping CDSs with different gene ids', 37),\n",
       "             ('Cas9 targets with unique location', 3665000),\n",
       "             ('Cas9 targets pass MM flt', 858176),\n",
       "             ('Cas9 targets pass GC flt', 3016380),\n",
       "             ('Cas9 targets pass Eff flt', 1946078),\n",
       "             ('Cpf1 targets with unique location', 2126127),\n",
       "             ('Cpf1 targets pass MM flt', 1695992),\n",
       "             ('Cpf1 targets pass GC flt', 1249707),\n",
       "             ('Cpf1 targets pass Eff flt', 716463),\n",
       "             ('Cas9 targets hit CDS', 2632337),\n",
       "             ('Cas9 targets pass all fileters', 379147),\n",
       "             ('Cpf1 targets hit CDS', 977085),\n",
       "             ('Cpf1 targets pass all fileters', 265419),\n",
       "             ('number of samples', 89),\n",
       "             ('vcf variants', 178535324),\n",
       "             ('variants', 161886064),\n",
       "             ('nucleotide diversity in coding regions', 0.013172444056535816)])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rank</th>\n",
       "      <th>Target sequence</th>\n",
       "      <th>Genomic location</th>\n",
       "      <th>Strand</th>\n",
       "      <th>GC content (%)</th>\n",
       "      <th>Self-complementarity</th>\n",
       "      <th>Efficiency</th>\n",
       "      <th>MM0</th>\n",
       "      <th>MM1</th>\n",
       "      <th>MM2</th>\n",
       "      <th>MM3</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>cutpos</th>\n",
       "      <th>cutneg</th>\n",
       "      <th>flag unique location</th>\n",
       "      <th>MM3_num</th>\n",
       "      <th>sumMM</th>\n",
       "      <th>flt MM</th>\n",
       "      <th>flt GC</th>\n",
       "      <th>flt Eff</th>\n",
       "      <th>idx</th>\n",
       "      <th>flt hitsCDS</th>\n",
       "      <th>vpos</th>\n",
       "      <th>nv</th>\n",
       "      <th>refAF</th>\n",
       "      <th>p_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>138</td>\n",
       "      <td>TTTTTTGATGTTTCATGTAGATGGCTCG</td>\n",
       "      <td>1:31480</td>\n",
       "      <td>+</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAEL012102-RB</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>31480</td>\n",
       "      <td>31503</td>\n",
       "      <td>31507</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>[31481, 31485, 31486, 31487, 31493, 31495]</td>\n",
       "      <td>6.0</td>\n",
       "      <td>[0.9943820224719101, 0.9887640449438202, 0.971...</td>\n",
       "      <td>0.849814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>134</td>\n",
       "      <td>TTTTTGATGTTTCATGTAGATGGCTCGC</td>\n",
       "      <td>1:31481</td>\n",
       "      <td>+</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAEL012102-RB</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>31481</td>\n",
       "      <td>31504</td>\n",
       "      <td>31508</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>[31481, 31485, 31486, 31487, 31493, 31495]</td>\n",
       "      <td>6.0</td>\n",
       "      <td>[0.9943820224719101, 0.9887640449438202, 0.971...</td>\n",
       "      <td>0.849814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>128</td>\n",
       "      <td>TTTTGATGTTTCATGTAGATGGCTCGCC</td>\n",
       "      <td>1:31482</td>\n",
       "      <td>+</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAEL012102-RB</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>31482</td>\n",
       "      <td>31505</td>\n",
       "      <td>31509</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>[31485, 31486, 31487, 31493, 31495]</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[0.9887640449438202, 0.9719101123595506, 0.904...</td>\n",
       "      <td>0.854615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>81</td>\n",
       "      <td>TTTGATGTTTCATGTAGATGGCTCGCCA</td>\n",
       "      <td>1:31483</td>\n",
       "      <td>+</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAEL012102-RB</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>31483</td>\n",
       "      <td>31506</td>\n",
       "      <td>31510</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>[31485, 31486, 31487, 31493, 31495]</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[0.9887640449438202, 0.9719101123595506, 0.904...</td>\n",
       "      <td>0.854615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17</td>\n",
       "      <td>TTTCATGTAGATGGCTCGCCATCTTAGA</td>\n",
       "      <td>1:31490</td>\n",
       "      <td>+</td>\n",
       "      <td>46</td>\n",
       "      <td>3</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAEL012102-RA</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>31490</td>\n",
       "      <td>31513</td>\n",
       "      <td>31517</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>[31493, 31495]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[0.9887640449438202, 0.9943820224719101]</td>\n",
       "      <td>0.983209</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Rank               Target sequence Genomic location Strand  GC content (%)  \\\n",
       "0   138  TTTTTTGATGTTTCATGTAGATGGCTCG          1:31480      +              42   \n",
       "1   134  TTTTTGATGTTTCATGTAGATGGCTCGC          1:31481      +              46   \n",
       "2   128  TTTTGATGTTTCATGTAGATGGCTCGCC          1:31482      +              50   \n",
       "3    81  TTTGATGTTTCATGTAGATGGCTCGCCA          1:31483      +              46   \n",
       "4    17  TTTCATGTAGATGGCTCGCCATCTTAGA          1:31490      +              46   \n",
       "\n",
       "   Self-complementarity  Efficiency  MM0  MM1  MM2 MM3  transcript_id gene_id  \\\n",
       "0                     0           4    0    0    0   0  AAEL012102-RB    None   \n",
       "1                     0           4    0    0    0   0  AAEL012102-RB    None   \n",
       "2                     0           7    0    0    0   0  AAEL012102-RB    None   \n",
       "3                     1          41    0    0    0   0  AAEL012102-RB    None   \n",
       "4                     3          75    0    0    0   0  AAEL012102-RA    None   \n",
       "\n",
       "  chrom    pos cutpos cutneg  flag unique location  MM3_num  sumMM  flt MM  \\\n",
       "0     1  31480  31503  31507                  True      0.0    0.0    True   \n",
       "1     1  31481  31504  31508                  True      0.0    0.0    True   \n",
       "2     1  31482  31505  31509                  True      0.0    0.0    True   \n",
       "3     1  31483  31506  31510                  True      0.0    0.0    True   \n",
       "4     1  31490  31513  31517                  True      0.0    0.0    True   \n",
       "\n",
       "   flt GC  flt Eff  idx  flt hitsCDS  \\\n",
       "0    True    False    0        False   \n",
       "1    True    False    1        False   \n",
       "2    True    False    2        False   \n",
       "3    True    False    3        False   \n",
       "4    True     True    4        False   \n",
       "\n",
       "                                         vpos   nv  \\\n",
       "0  [31481, 31485, 31486, 31487, 31493, 31495]  6.0   \n",
       "1  [31481, 31485, 31486, 31487, 31493, 31495]  6.0   \n",
       "2         [31485, 31486, 31487, 31493, 31495]  5.0   \n",
       "3         [31485, 31486, 31487, 31493, 31495]  5.0   \n",
       "4                              [31493, 31495]  2.0   \n",
       "\n",
       "                                               refAF     p_ref  \n",
       "0  [0.9943820224719101, 0.9887640449438202, 0.971...  0.849814  \n",
       "1  [0.9943820224719101, 0.9887640449438202, 0.971...  0.849814  \n",
       "2  [0.9887640449438202, 0.9719101123595506, 0.904...  0.854615  \n",
       "3  [0.9887640449438202, 0.9719101123595506, 0.904...  0.854615  \n",
       "4           [0.9887640449438202, 0.9943820224719101]  0.983209  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display_all_cols(targets['Cpf1'].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 44.2 s, sys: 2.26 s, total: 46.4 s\n",
      "Wall time: 48.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Saving targets dataframe\n",
    "with open(f'{SETNAME}_CDS_targets.pickle', 'wb') as fh:\n",
    "    pickle.dump(targets, fh)\n",
    "with open(f'{SETNAME}_CDS_results_dict.pickle', 'wb') as fh:\n",
    "    pickle.dump(results_dict, fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#### Cas9\n",
      "#### Cpf1\n",
      "CPU times: user 5.34 s, sys: 236 ms, total: 5.58 s\n",
      "Wall time: 6.48 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# saving subset of potential targets (passing all filters except population variation)\n",
    "# this is much smaller than all the unique targets, so speeds up the later analysis, figure generation, ect.\n",
    "potential_targets = OrderedDict()\n",
    "\n",
    "for mode in targets.keys():\n",
    "    print('####', mode)\n",
    "    d = targets[mode]\n",
    "    potential_targets[mode] = d[(d['flt MM']==True) & (d['flt GC']==True) & (d['flt Eff']==True) & (d['flt hitsCDS']==True)].copy()\n",
    "\n",
    "# Saving the results and potential_targets dataframe (smaller and used for analysis)\n",
    "with open(f'{SETNAME}_potential_targets.pickle', 'wb') as fh:\n",
    "    pickle.dump((potential_targets, results_dict), fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
